{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Things to calculate:\n",
    "total/mt+/mt-\n",
    " - Pi N\n",
    " - Pi S\n",
    " - Theta W\n",
    " - S\n",
    " - TajD\n",
    "\n",
    "between\n",
    " - Fst between alleles\n",
    " - fixed/shared/private alleles\n",
    "total\n",
    " - GC content at synonymous sites\n",
    " \n",
    "Codon stats\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "from ness_fasta import ness_fasta\n",
    "from Bio import SeqIO, AlignIO\n",
    "from collections import Counter\n",
    "from ness_vcf import SFS\n",
    "import glob\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_column(position, aln, samples=None):\n",
    "    #sample_IDs = [sample.id.split(\"|\")[0] for sample in aln]\n",
    "    sample_IDs = [s.id for s in aln]\n",
    "    column = aln[:,position]\n",
    "    col = dict(zip(sample_IDs, [i for i in column]))\n",
    "    if samples != None:\n",
    "        col_filtered ={}\n",
    "        for s in col:\n",
    "            if s in samples:\n",
    "                col_filtered[s]=col[s]\n",
    "        col=col_filtered\n",
    "    return col\n",
    "\n",
    "def alignment2sfs(aln, min_called, samples=None, start=None, end=None):\n",
    "    if start== None: start =0\n",
    "    if end == None:end = len(aln[0].seq)\n",
    "    if samples == None:\n",
    "        samples=[s.id for s in aln]\n",
    "    sfs = SFS([0]*(len(samples)+1))\n",
    "    for position in range(start, end):\n",
    "        column = get_column(position, aln, samples)\n",
    "        bases=column.values()\n",
    "        callable_bases = [column[sampleID] for sampleID in samples if column[sampleID] in 'AaCcGgTt']\n",
    "        if len(callable_bases) >=min_called: # if callable\n",
    "            base_counts = Counter(callable_bases)\n",
    "            if len(base_counts) == 1: # invariant\n",
    "                sfs.add(0,len(samples))\n",
    "            elif len(base_counts) >2:  # triallelic\n",
    "                continue\n",
    "            elif len(base_counts)==2:\n",
    "                #print(len(base_counts), callable_bases)\n",
    "                AF = Counter(bases)[list(Counter(bases).keys())[1]]/len(bases)\n",
    "                sfs.add(AF,len(samples))\n",
    "    return sfs\n",
    "\n",
    "def concatenate_SFSs(SFSs):\n",
    "    concatenate_sfs =[]\n",
    "    for s in SFSs:\n",
    "        for i in range(len(s)):\n",
    "            try:combined_sfs[i] += s[i]\n",
    "            except IndexError:\n",
    "                \"You can't combine SFSs with different numbers of samples\"\n",
    "                return None\n",
    "    return combined_sfs\n",
    "\n",
    "gen_code_dict = { 'ttt': 'F', 'ttc': 'F', 'tta': 'L', 'ttg': 'L', \\\n",
    "                'tct': 'S', 'tcc': 'S', 'tca': 'S', 'tcg': 'S', \\\n",
    "                'tat': 'Y', 'tac': 'Y', 'taa': 'X', 'tag': 'X', \\\n",
    "                'tgt': 'C', 'tgc': 'C', 'tga': 'X', 'tgg': 'W', \\\n",
    "                'ctt': 'L', 'ctc': 'L', 'cta': 'L', 'ctg': 'L', \\\n",
    "                'cct': 'P', 'ccc': 'P', 'cca': 'P', 'ccg': 'P', \\\n",
    "                'cat': 'H', 'cac': 'H', 'caa': 'Q', 'cag': 'Q', \\\n",
    "                'cgt': 'R', 'cgc': 'R', 'cga': 'R', 'cgg': 'R', \\\n",
    "                'att': 'I', 'atc': 'I', 'ata': 'I', 'atg': 'M', \\\n",
    "                'act': 'T', 'acc': 'T', 'aca': 'T', 'acg': 'T', \\\n",
    "                'aat': 'N', 'aac': 'N', 'aaa': 'K', 'aag': 'K', \\\n",
    "                'agt': 'S', 'agc': 'S', 'aga': 'R', 'agg': 'R', \\\n",
    "                'gtt': 'V', 'gtc': 'V', 'gta': 'V', 'gtg': 'V', \\\n",
    "                'gct': 'A', 'gcc': 'A', 'gca': 'A', 'gcg': 'A', \\\n",
    "                'gat': 'D', 'gac': 'D', 'gaa': 'E', 'gag': 'E', \\\n",
    "                'ggt': 'G', 'ggc': 'G', 'gga': 'G', 'ggg': 'G'}\n",
    "\n",
    "degen_dict = {'ttt': '002', 'ttc': '002', 'tta': '202', 'ttg': '202', \\\n",
    "              'tct': '004', 'tcc': '004', 'tca': '004', 'tcg': '004', \\\n",
    "              'tat': '002', 'tac': '002', 'taa': '022', 'tag': '002', \\\n",
    "              'tgt': '002', 'tgc': '002', 'tga': '020', 'tgg': '000', \\\n",
    "              'ctt': '004', 'ctc': '004', 'cta': '204', 'ctg': '204', \\\n",
    "              'cct': '004', 'ccc': '004', 'cca': '004', 'ccg': '004', \\\n",
    "              'cat': '002', 'cac': '002', 'caa': '002', 'cag': '002', \\\n",
    "              'cgt': '004', 'cgc': '004', 'cga': '204', 'cgg': '204', \\\n",
    "              'att': '003', 'atc': '003', 'ata': '003', 'atg': '000', \\\n",
    "              'act': '004', 'acc': '004', 'aca': '004', 'acg': '004', \\\n",
    "              'aat': '002', 'aac': '002', 'aaa': '002', 'aag': '002', \\\n",
    "              'agt': '002', 'agc': '002', 'aga': '202', 'agg': '202', \\\n",
    "              'gtt': '004', 'gtc': '004', 'gta': '004', 'gtg': '004', \\\n",
    "              'gct': '004', 'gcc': '004', 'gca': '004', 'gcg': '004', \\\n",
    "              'gat': '002', 'gac': '002', 'gaa': '002', 'gag': '002', \\\n",
    "              'ggt': '004', 'ggc': '004', 'gga': '004', 'ggg': '004'}\n",
    "\n",
    "\n",
    "def special_consensus(aln):\n",
    "    consensus = ''\n",
    "    for p in range(len(aln[0].seq)):\n",
    "        called_bases = [b for b in aln[:,p] if b in \"ACTG\"]\n",
    "        try:\n",
    "            most_common_nucleotide = list(Counter(called_bases).keys())[0]\n",
    "            consensus+=most_common_nucleotide\n",
    "        except IndexError:\n",
    "            consensus+='N'\n",
    "    return consensus\n",
    "        \n",
    "def degenerate(seq):\n",
    "    degen_seq = ''\n",
    "    for idx in range(0, len(aln[0])-2,3):    \n",
    "        codon = seq[idx:idx+3]\n",
    "        try:\n",
    "            degen  = degen_dict[codon.lower()]\n",
    "        except KeyError:\n",
    "            degen = 'nnn'\n",
    "        degen_seq = degen_seq+degen\n",
    "    return degen_seq\n",
    "\n",
    "def alignment2sfs_w_degen(aln, min_called, degen_map, samples=None, start=None, end=None):\n",
    "    if start== None: start =0\n",
    "    if end == None:end = len(aln[0].seq)\n",
    "    if samples == None:\n",
    "        samples=[s.id for s in aln]\n",
    "    sfs0 = SFS([0]*(len(samples)+1))\n",
    "    sfs4 = SFS([0]*(len(samples)+1))\n",
    "    sfstotal = SFS([0]*(len(samples)+1))\n",
    "    for position in range(start, end):\n",
    "        degeneracy = degen_map[position]\n",
    "        column = get_column(position, aln, samples)\n",
    "        bases=column.values()\n",
    "        callable_bases = [column[sampleID] for sampleID in samples if column[sampleID] in 'AaCcGgTt']\n",
    "        if len(callable_bases) >=min_called: # if callable\n",
    "            base_counts = Counter(callable_bases)\n",
    "            if len(base_counts) == 1: # invariant\n",
    "                sfstotal.add(0,len(samples))\n",
    "                if degeneracy=='0':sfs0.add(0,len(samples))\n",
    "                if degeneracy=='4':sfs4.add(0,len(samples))\n",
    "            elif len(base_counts) >2:  # triallelic\n",
    "                continue\n",
    "            elif len(base_counts)==2:\n",
    "                #print(len(base_counts), callable_bases)\n",
    "                AF = Counter(bases)[list(Counter(bases).keys())[1]]/len(bases)\n",
    "                sfstotal.add(AF,len(samples))\n",
    "                if degeneracy=='0':sfs0.add(AF,len(samples))\n",
    "                if degeneracy=='4':sfs4.add(AF,len(samples))\n",
    "\n",
    "    return [sfstotal, sfs0, sfs4]\n",
    "\n",
    "def population_poly(aln, min_called, degen_map, group1, group2, start=None, end=None):\n",
    "    ## Calculate stats between groups:\n",
    "    #     SNP distribution stats:\n",
    "    #     private = polymorphic in one group and not the other\n",
    "    #     fixed = groups are monomorphic within but different\n",
    "    #     shared = polymorphic in both\n",
    "    ## Fst  - differentiation\n",
    "    #     Fst = 1 - [ mean(2*p_1*q_1, 2*p_2*q_2) / (2*p_total*q_total) ]\n",
    "    if start== None: start =0\n",
    "    if end == None:end = len(aln[0].seq)\n",
    "    group1_samples=[s.id for s in aln if s.id in group1]\n",
    "    group2_samples=[s.id for s in aln if s.id in group2]\n",
    "    all_samples=group1_samples+group2_samples\n",
    "    fst = [0, 0]\n",
    "    private1, private2,fixed,shared = 0,0,0,0 \n",
    "    for position in range(start, end):\n",
    "        degeneracy = degen_map[position]\n",
    "        column = get_column(position, aln, all_samples)\n",
    "        bases=column.values()\n",
    "        all_callable_bases = [column[sampleID] for sampleID in all_samples if column[sampleID] in 'AaCcGgTt']\n",
    "        group1_callable_bases = [column[sampleID] for sampleID in group1_samples if column[sampleID] in 'AaCcGgTt']\n",
    "        group2_callable_bases = [column[sampleID] for sampleID in group2_samples if column[sampleID] in 'AaCcGgTt']\n",
    "        if len(all_callable_bases) >=min_called: # if callable\n",
    "            base_counts_all = Counter(all_callable_bases)\n",
    "            base_counts_1 = Counter(group1_callable_bases)\n",
    "            base_counts_2 = Counter(group2_callable_bases)\n",
    "            #if this is monomorphic - move on!\n",
    "            if len(base_counts_all) == 1:# invariant\n",
    "                # Add invariant counts to all or do nothing  \n",
    "                pass\n",
    "            elif len(base_counts_all) >2:  # triallelic\n",
    "                continue\n",
    "            elif len(base_counts_all)==2:\n",
    "                p_1 = base_counts_1[list(base_counts_1.keys())[0]]/len(group1_callable_bases)\n",
    "                p_2 = base_counts_2[list(base_counts_2.keys())[0]]/len(group2_callable_bases)\n",
    "                p_all = base_counts_all[list(base_counts_all.keys())[0]]/len(all_callable_bases)\n",
    "                pq_1 = 2*p_1*(1-p_1)\n",
    "                pq_2 = 2*p_2*(1-p_2)\n",
    "                pq_within=(pq_1+pq_2)/2\n",
    "                pq_all = 2*p_all*(1-p_all)\n",
    "                fst_i = 1-(pq_within/pq_all)\n",
    "                fst[0]+=fst_i #running sum of fst\n",
    "                fst[1]+=1 # total sites in there\n",
    "                if p_1 == 1.0 and 0<p_2<1:\n",
    "                    private2+=1\n",
    "                elif p_2 == 1.0 and 0<p_1<1:\n",
    "                    private1+=1\n",
    "#                     print(group1_callable_bases, group2_callable_bases, p_1, p_2)\n",
    "                if 0<p_1<1 and 0<p_2<1:\n",
    "                    shared+=1\n",
    "                if p_1 ==1.0 and p_2 ==1.0 and 0<p_all<1:\n",
    "                    fixed +=1                    \n",
    " #########################################################################################################\n",
    "    try:fst =  fst[0]/fst[1]\n",
    "    except ZeroDivisionError:\n",
    "        fst = 'N/A'\n",
    "    results = { 'private1': private1,\n",
    "                'private2': private2,\n",
    "                'fixed': fixed,\n",
    "                'shared': shared,\n",
    "                'fst' :fst}\n",
    "    return results\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Target Genotype Lists\n",
    "mtPlus_Genotypes = \"CC2936 CC2937 CC3060 CC3064 CC3065 CC3068 CC3071 CC3076 CC3086\".split()\n",
    "mtMinus_Genotypes = \"CC2935 CC2938 CC3059 CC3061 CC3062 CC3063 CC3073 CC3075 CC3079 CC3084\".split()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EIF5Bb.quebec_all.aln.fasta EIF5Bb quebec_all.al\n",
      "MT0829.quebec_all.aln.fasta MT0829 quebec_all.al\n",
      "MT0828.quebec_all.aln.fasta MT0828 quebec_all.al\n",
      "NIC7.quebec_all.aln (modified).fasta NIC7 quebec_all.aln (modified)\n",
      "RFC4.quebec_all.aln.fasta RFC4 quebec_all.al\n",
      "PKY1.quebec_all.aln.fasta PKY1 quebec_all.al\n",
      "LPS1.quebec_all.aln (modified).fasta LPS1 quebec_all.aln (modified)\n",
      "522915.quebec_all.aln.fasta 522915 quebec_all.al\n",
      "522922.quebec_all.aln.fasta 522922 quebec_all.al\n",
      "294752.quebec_all.aln.fasta 294752 quebec_all.al\n",
      "522918.quebec_all.aln.fasta 522918 quebec_all.al\n",
      "DLA3.quebec_all.aln.fasta DLA3 quebec_all.al\n",
      "DRG1.quebec_all.aln.fasta DRG1 quebec_all.al\n",
      "196073.quebec_all.aln.fasta 196073 quebec_all.al\n",
      "HDH1.quebec_all.aln.fasta HDH1 quebec_all.al\n",
      "NMDA1.quebec_all.aln.fasta NMDA1 quebec_all.al\n",
      "GCSH.quebec_all.aln.fasta GCSH quebec_all.al\n",
      "PSF2.quebec_all.aln.fasta PSF2 quebec_all.al\n",
      "344092.quebec_all.aln.fasta 344092 quebec_all.al\n",
      "196063.quebec_all.aln.fasta 196063 quebec_all.al\n",
      "522914.quebec_all.aln.fasta 522914 quebec_all.al\n",
      "182389.quebec_all.aln.fasta 182389 quebec_all.al\n",
      "CGL70.quebec_all.aln.fasta CGL70 quebec_all.al\n",
      "HRGP1.quebec_all.aln- partially realigned.fasta HRGP1 quebec_all.aln- partially realigned\n",
      "PTC1.quebec_all.aln.fasta PTC1 quebec_all.al\n",
      "FUM1.quebec_all.aln.fasta FUM1 quebec_all.al\n",
      "294742.quebec_all.aln.fasta 294742 quebec_all.al\n",
      "SPL2.quebec_all.aln.fasta SPL2 quebec_all.al\n",
      "TOC34.quebec_all.aln.fasta TOC34 quebec_all.al\n",
      "PR46b.quebec_all.aln Copy - realigned (modified).fasta PR46b quebec_all.aln Copy - realigned (modified)\n",
      "FBX9.quebec_all.aln.fasta FBX9 quebec_all.al\n",
      "ALB3.quebec_all.aln.fasta ALB3 quebec_all.al\n",
      "UTP1.quebec_all.aln.fasta UTP1 quebec_all.al\n",
      "522919.quebec_all.aln.fasta 522919 quebec_all.al\n",
      "MT0796.quebec_all.aln.fasta MT0796 quebec_all.al\n",
      "PDK1.quebec_all.aln.fasta PDK1 quebec_all.al\n",
      "182392.quebec_all.aln.fasta 182392 quebec_all.al\n",
      "182390.quebec_all.aln.fasta 182390 quebec_all.al\n",
      "97782.quebec_all.aln.fasta 97782 quebec_all.al\n",
      "522872.quebec_all.aln (modified).fasta 522872 quebec_all.aln (modified)\n",
      "UBCH1.quebec_all.aln.fasta UBCH1 quebec_all.al\n",
      "SPP3.quebec_all.aln Copy - realigned.fasta SPP3 quebec_all.aln Copy - realigned\n",
      "SPS1.quebec_all.aln.fasta SPS1 quebec_all.al\n",
      "182394.quebec_all.aln.fasta 182394 quebec_all.al\n",
      "THI10.quebec_all.aln.fasta THI10 quebec_all.al\n",
      "MADS2.quebec_all.aln.fasta MADS2 quebec_all.al\n",
      "CGLD28.quebec_all.aln.fasta CGLD28 quebec_all.al\n",
      "LEU1S.quebec_all.aln.fasta LEU1S quebec_all.al\n",
      "522875.quebec_all.aln.fasta 522875 quebec_all.al\n",
      "SAD1.quebec_all.aln.fasta SAD1 quebec_all.al\n",
      "294687.quebec_all.aln.fasta 294687 quebec_all.al\n",
      "51 0 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51 51\n"
     ]
    }
   ],
   "source": [
    "#### STATS TO COLLECT for each gene\n",
    "genes = []\n",
    "ness_ids = []\n",
    "mtPlus_coords = []\n",
    "mtMinus_coords = []\n",
    "alleles_ALL = []\n",
    "sites_ALL = []\n",
    "invariant_ALL = []\n",
    "variant_ALL = []\n",
    "theta_pi_0fold_ALL = []\n",
    "theta_pi_4fold_ALL = []\n",
    "theta_pi_total_ALL = []\n",
    "theta_w_0fold_ALL = []\n",
    "theta_w_4fold_ALL = []\n",
    "theta_w_total_ALL = []\n",
    "sfs_4fold_ALL = []\n",
    "sfs_0fold_ALL = []\n",
    "sfs_t_ALL = []\n",
    "alleles_MTPLUS = []\n",
    "sites_MTPLUS = []\n",
    "invariant_MTPLUS = []\n",
    "variant_MTPLUS = []\n",
    "theta_pi_0fold_MTPLUS = []\n",
    "theta_pi_4fold_MTPLUS = []\n",
    "theta_pi_total_MTPLUS = []\n",
    "theta_w_0fold_MTPLUS = []\n",
    "theta_w_4fold_MTPLUS = []\n",
    "theta_w_total_MTPLUS = []\n",
    "sfs_4fold_MTPLUS = []\n",
    "sfs_0fold_MTPLUS = []\n",
    "sfs_t_MTPLUS = []\n",
    "alleles_MTMINUS = []\n",
    "sites_MTMINUS = []\n",
    "invariant_MTMINUS = []\n",
    "variant_MTMINUS = []\n",
    "theta_pi_0fold_MTMINUS = []\n",
    "theta_pi_4fold_MTMINUS = []\n",
    "theta_pi_total_MTMINUS = []\n",
    "theta_w_0fold_MTMINUS = []\n",
    "theta_w_4fold_MTMINUS = []\n",
    "theta_w_total_MTMINUS = []\n",
    "sfs_4fold_MTMINUS = []\n",
    "sfs_0fold_MTMINUS = []\n",
    "sfs_t_MTMINUS = []\n",
    "private_MTPLUS = []\n",
    "private_MTMINUS = []\n",
    "shared = []\n",
    "fixed = []\n",
    "Fst = []\n",
    "GC4 = []\n",
    "\n",
    "# Loop over all shared coding gametologs\n",
    "alignment_files = glob.glob('../TranslationAligner/curated/*.fasta')\n",
    "for a in alignment_files[:]:\n",
    "    gene = a.split(\"/\")[-1].split(\".\")[0]\n",
    "    genes.append(gene)\n",
    "    ness_id = a.split(\"/\")[-1].lstrip(gene+\".\").rstrip(\".quebec_mtPlus.fasta\").rstrip(\".quebec_mtMinus.fasta\")\n",
    "    print(a.split(\"/\")[-1],gene, ness_id)\n",
    "    aln = AlignIO.read(open(a), 'fasta')\n",
    "    mtPlus_sampleIDs  = [s.id for s in aln if s.id.split(\"|\")[0] in mtPlus_Genotypes]\n",
    "    mtMinus_sampleIDs = [s.id for s in aln if s.id.split(\"|\")[0] in mtMinus_Genotypes]\n",
    "    mtPlus_coords.append(mtPlus_sampleIDs[0].split(\"|\")[-1]) ## based on this kind of name e.g. \"CC2935|mtMinus:105065-114642\"\n",
    "    mtMinus_coords.append(mtMinus_sampleIDs[0].split(\"|\")[-1])\n",
    "    degen_map = degenerate(special_consensus(aln))\n",
    "    sfst,sfs0,sfs4 = alignment2sfs_w_degen(aln, 12, degen_map, samples=None, start=None, end=None)\n",
    "    ##### record ALL results\n",
    "    alleles_ALL.append(sfst.alleles)\n",
    "    sites_ALL.append(sfst.sites())\n",
    "    invariant_ALL.append(sfst.invariant())\n",
    "    variant_ALL.append(sfst.variant())\n",
    "    theta_pi_0fold_ALL.append(sfs0.theta_pi())\n",
    "    theta_pi_4fold_ALL.append(sfs4.theta_pi())\n",
    "    theta_pi_total_ALL.append(sfst.theta_pi())\n",
    "    theta_w_0fold_ALL.append(sfs0.theta_w())\n",
    "    theta_w_4fold_ALL.append(sfs4.theta_w())\n",
    "    theta_w_total_ALL.append(sfst.theta_w())\n",
    "    sfs_4fold_ALL.append(sfs4.sfs)\n",
    "    sfs_0fold_ALL.append(sfs0.sfs)\n",
    "    sfs_t_ALL.append(sfst.sfs)\n",
    "    ###################################\n",
    "    ## MTPLUS\n",
    "    sfst,sfs0,sfs4 = alignment2sfs_w_degen(aln, 8, degen_map, samples=mtPlus_sampleIDs, start=None, end=None)\n",
    "    ##### record ALL results\n",
    "    alleles_MTPLUS.append(sfst.alleles)\n",
    "    sites_MTPLUS.append(sfst.sites())\n",
    "    invariant_MTPLUS.append(sfst.invariant())\n",
    "    variant_MTPLUS.append(sfst.variant())\n",
    "    theta_pi_0fold_MTPLUS.append(sfs0.theta_pi())\n",
    "    theta_pi_4fold_MTPLUS.append(sfs4.theta_pi())\n",
    "    theta_pi_total_MTPLUS.append(sfst.theta_pi())\n",
    "    theta_w_0fold_MTPLUS.append(sfs0.theta_w())\n",
    "    theta_w_4fold_MTPLUS.append(sfs4.theta_w())\n",
    "    theta_w_total_MTPLUS.append(sfst.theta_w())\n",
    "    sfs_4fold_MTPLUS.append(sfs4.sfs)\n",
    "    sfs_0fold_MTPLUS.append(sfs0.sfs)\n",
    "    sfs_t_MTPLUS.append(sfst.sfs)   \n",
    "    ###################################\n",
    "    ## MTMINUS\n",
    "    sfst,sfs0,sfs4 = alignment2sfs_w_degen(aln, 8, degen_map, samples=mtMinus_sampleIDs, start=None, end=None)\n",
    "    ##### record ALL results\n",
    "    alleles_MTMINUS.append(sfst.alleles)\n",
    "    sites_MTMINUS.append(sfst.sites())\n",
    "    invariant_MTMINUS.append(sfst.invariant())\n",
    "    variant_MTMINUS.append(sfst.variant())\n",
    "    theta_pi_0fold_MTMINUS.append(sfs0.theta_pi())\n",
    "    theta_pi_4fold_MTMINUS.append(sfs4.theta_pi())\n",
    "    theta_pi_total_MTMINUS.append(sfst.theta_pi())\n",
    "    theta_w_0fold_MTMINUS.append(sfs0.theta_w())\n",
    "    theta_w_4fold_MTMINUS.append(sfs4.theta_w())\n",
    "    theta_w_total_MTMINUS.append(sfst.theta_w())\n",
    "    sfs_4fold_MTMINUS.append(sfs4.sfs)\n",
    "    sfs_0fold_MTMINUS.append(sfs0.sfs)\n",
    "    sfs_t_MTMINUS.append(sfst.sfs)   \n",
    "    ###################################\n",
    "    ## population stats\n",
    "    pop_stats= population_poly(aln, 12, degen_map, mtPlus_sampleIDs, mtMinus_sampleIDs, start=None, end=None)\n",
    "    private_MTPLUS.append(  pop_stats['private1'])\n",
    "    private_MTMINUS.append( pop_stats['private2'])\n",
    "    shared.append(pop_stats['shared'])\n",
    "    fixed.append(pop_stats['fixed'])\n",
    "    Fst.append(pop_stats['fst'])\n",
    "    #######\n",
    "    ## GC_4fold\n",
    "    GC4.append(len([1 for i,j in zip(special_consensus(aln),degen_map) if i in 'GCgc' and j =='4'])/\n",
    "               len([1 for i,j in zip(special_consensus(aln),degen_map) if i in 'AaTtGCgc' and j =='4']))\n",
    "\n",
    "#########################################################\n",
    "#### DF the results #####################################\n",
    "#########################################################\n",
    "print(len(genes), len(ness_ids), len(mtPlus_coords), len(mtMinus_coords), len(alleles_ALL), len(sites_ALL), \n",
    "      len(invariant_ALL), len(variant_ALL), len(theta_pi_0fold_ALL), len(theta_pi_4fold_ALL), len(theta_pi_total_ALL), \n",
    "      len(theta_w_0fold_ALL), len(theta_w_4fold_ALL), len(theta_w_total_ALL), len(sfs_4fold_ALL), len(sfs_0fold_ALL), \n",
    "      len(sfs_t_ALL), len(alleles_MTPLUS), len(sites_MTPLUS), len(invariant_MTPLUS), len(variant_MTPLUS), \n",
    "      len(theta_pi_0fold_MTPLUS), len(theta_pi_4fold_MTPLUS), len(theta_pi_total_MTPLUS), len(theta_w_0fold_MTPLUS), \n",
    "      len(theta_w_4fold_MTPLUS), len(theta_w_total_MTPLUS), len(sfs_4fold_MTPLUS), len(sfs_0fold_MTPLUS), len(sfs_t_MTPLUS), \n",
    "      len(alleles_MTMINUS), len(sites_MTMINUS), len(invariant_MTMINUS), len(variant_MTMINUS), len(theta_pi_0fold_MTMINUS), \n",
    "      len(theta_pi_4fold_MTMINUS), len(theta_pi_total_MTMINUS), len(theta_w_0fold_MTMINUS), len(theta_w_4fold_MTMINUS), \n",
    "      len(theta_w_total_MTMINUS), len(sfs_4fold_MTMINUS), len(sfs_0fold_MTMINUS), len(sfs_t_MTMINUS), \n",
    "      len(private_MTPLUS), len(private_MTMINUS), len(shared), len(fixed), len(Fst), len(GC4))\n",
    "\n",
    "\n",
    "shared_results = pd.DataFrame(\n",
    "    {'gene' : genes,\n",
    "#     'ness_id': ness_ids, # breaking df\n",
    "     'mtPlus_coords' : mtPlus_coords,\n",
    "     'mtMinus_coords' : mtMinus_coords,\n",
    "     'alleles_ALL' : alleles_ALL, \n",
    "     'sites_ALL' : sites_ALL, \n",
    "     'invariant_ALL' : invariant_ALL, \n",
    "     'variant_ALL' : variant_ALL, \n",
    "     'theta_pi_0fold_ALL' : theta_pi_0fold_ALL, \n",
    "     'theta_pi_4fold_ALL' : theta_pi_4fold_ALL, \n",
    "     'theta_pi_total_ALL' : theta_pi_total_ALL, \n",
    "     'theta_w_0fold_ALL' : theta_w_0fold_ALL, \n",
    "     'theta_w_4fold_ALL' : theta_w_4fold_ALL, \n",
    "     'theta_w_total_ALL' : theta_w_total_ALL, \n",
    "     'sfs_4fold_ALL' : sfs_4fold_ALL, \n",
    "     'sfs_0fold_ALL' : sfs_0fold_ALL, \n",
    "     'sfs_t_ALL' : sfs_t_ALL,    \n",
    "     'alleles_MTPLUS' : alleles_MTPLUS, \n",
    "     'sites_MTPLUS' : sites_MTPLUS, \n",
    "     'invariant_MTPLUS' : invariant_MTPLUS, \n",
    "     'variant_MTPLUS' : variant_MTPLUS, \n",
    "     'theta_pi_0fold_MTPLUS' : theta_pi_0fold_MTPLUS, \n",
    "     'theta_pi_4fold_MTPLUS' : theta_pi_4fold_MTPLUS, \n",
    "     'theta_pi_total_MTPLUS' : theta_pi_total_MTPLUS, \n",
    "     'theta_w_0fold_MTPLUS' : theta_w_0fold_MTPLUS, \n",
    "     'theta_w_4fold_MTPLUS' : theta_w_4fold_MTPLUS, \n",
    "     'theta_w_total_MTPLUS' : theta_w_total_MTPLUS, \n",
    "     'sfs_4fold_MTPLUS' : sfs_4fold_MTPLUS, \n",
    "     'sfs_0fold_MTPLUS' : sfs_0fold_MTPLUS, \n",
    "     'sfs_t_MTPLUS' : sfs_t_MTPLUS,         \n",
    "     'alleles_MTMINUS' : alleles_MTMINUS, \n",
    "     'sites_MTMINUS' : sites_MTMINUS, \n",
    "     'invariant_MTMINUS' : invariant_MTMINUS, \n",
    "     'variant_MTMINUS' : variant_MTMINUS, \n",
    "     'theta_pi_0fold_MTMINUS' : theta_pi_0fold_MTMINUS, \n",
    "     'theta_pi_4fold_MTMINUS' : theta_pi_4fold_MTMINUS, \n",
    "     'theta_pi_total_MTMINUS' : theta_pi_total_MTMINUS, \n",
    "     'theta_w_0fold_MTMINUS' : theta_w_0fold_MTMINUS, \n",
    "     'theta_w_4fold_MTMINUS' : theta_w_4fold_MTMINUS, \n",
    "     'theta_w_total_MTMINUS' : theta_w_total_MTMINUS, \n",
    "     'sfs_4fold_MTMINUS' : sfs_4fold_MTMINUS, \n",
    "     'sfs_0fold_MTMINUS' : sfs_0fold_MTMINUS, \n",
    "     'sfs_t_MTMINUS' : sfs_t_MTMINUS, \n",
    "     'private_MTPLUS' : private_MTPLUS, \n",
    "     'private_MTMINUS' : private_MTMINUS, \n",
    "     'shared' : shared, \n",
    "     'fixed' : fixed, \n",
    "     'Fst' : Fst,\n",
    "     'GC4' : GC4\n",
    "    })     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Fst</th>\n",
       "      <th>GC4</th>\n",
       "      <th>alleles_ALL</th>\n",
       "      <th>alleles_MTMINUS</th>\n",
       "      <th>alleles_MTPLUS</th>\n",
       "      <th>fixed</th>\n",
       "      <th>gene</th>\n",
       "      <th>invariant_ALL</th>\n",
       "      <th>invariant_MTMINUS</th>\n",
       "      <th>invariant_MTPLUS</th>\n",
       "      <th>...</th>\n",
       "      <th>theta_w_0fold_MTPLUS</th>\n",
       "      <th>theta_w_4fold_ALL</th>\n",
       "      <th>theta_w_4fold_MTMINUS</th>\n",
       "      <th>theta_w_4fold_MTPLUS</th>\n",
       "      <th>theta_w_total_ALL</th>\n",
       "      <th>theta_w_total_MTMINUS</th>\n",
       "      <th>theta_w_total_MTPLUS</th>\n",
       "      <th>variant_ALL</th>\n",
       "      <th>variant_MTMINUS</th>\n",
       "      <th>variant_MTPLUS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.034259</td>\n",
       "      <td>0.866279</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>EIF5Bb</td>\n",
       "      <td>929</td>\n",
       "      <td>1080</td>\n",
       "      <td>991</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.010146</td>\n",
       "      <td>0.008518</td>\n",
       "      <td>0.010012</td>\n",
       "      <td>0.006030</td>\n",
       "      <td>0.006111</td>\n",
       "      <td>0.005846</td>\n",
       "      <td>20</td>\n",
       "      <td>19</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.392467</td>\n",
       "      <td>0.809091</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>MT0829</td>\n",
       "      <td>1066</td>\n",
       "      <td>1075</td>\n",
       "      <td>1088</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002562</td>\n",
       "      <td>0.018290</td>\n",
       "      <td>0.019369</td>\n",
       "      <td>0.005017</td>\n",
       "      <td>0.007323</td>\n",
       "      <td>0.006139</td>\n",
       "      <td>0.003351</td>\n",
       "      <td>28</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.466787</td>\n",
       "      <td>0.976190</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>MT0828</td>\n",
       "      <td>147</td>\n",
       "      <td>77</td>\n",
       "      <td>169</td>\n",
       "      <td>...</td>\n",
       "      <td>0.010316</td>\n",
       "      <td>0.016830</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.017521</td>\n",
       "      <td>0.013005</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.010573</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.131916</td>\n",
       "      <td>0.854801</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>NIC7</td>\n",
       "      <td>1443</td>\n",
       "      <td>1716</td>\n",
       "      <td>1654</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000686</td>\n",
       "      <td>0.020373</td>\n",
       "      <td>0.016967</td>\n",
       "      <td>0.022688</td>\n",
       "      <td>0.007341</td>\n",
       "      <td>0.006670</td>\n",
       "      <td>0.008688</td>\n",
       "      <td>38</td>\n",
       "      <td>33</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.092847</td>\n",
       "      <td>0.798701</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>RFC4</td>\n",
       "      <td>641</td>\n",
       "      <td>724</td>\n",
       "      <td>651</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000918</td>\n",
       "      <td>0.014619</td>\n",
       "      <td>0.020658</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004395</td>\n",
       "      <td>0.005763</td>\n",
       "      <td>0.000564</td>\n",
       "      <td>10</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.660817</td>\n",
       "      <td>0.779241</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>64</td>\n",
       "      <td>PKY1</td>\n",
       "      <td>5275</td>\n",
       "      <td>6158</td>\n",
       "      <td>7876</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000722</td>\n",
       "      <td>0.010128</td>\n",
       "      <td>0.005515</td>\n",
       "      <td>0.001018</td>\n",
       "      <td>0.006882</td>\n",
       "      <td>0.003411</td>\n",
       "      <td>0.000839</td>\n",
       "      <td>130</td>\n",
       "      <td>60</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.583469</td>\n",
       "      <td>0.517949</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>LPS1</td>\n",
       "      <td>654</td>\n",
       "      <td>2257</td>\n",
       "      <td>1666</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000338</td>\n",
       "      <td>0.017882</td>\n",
       "      <td>0.010941</td>\n",
       "      <td>0.001256</td>\n",
       "      <td>0.008490</td>\n",
       "      <td>0.005853</td>\n",
       "      <td>0.000441</td>\n",
       "      <td>20</td>\n",
       "      <td>38</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.315672</td>\n",
       "      <td>0.718750</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>522915</td>\n",
       "      <td>556</td>\n",
       "      <td>564</td>\n",
       "      <td>559</td>\n",
       "      <td>...</td>\n",
       "      <td>0.018642</td>\n",
       "      <td>0.007036</td>\n",
       "      <td>0.008692</td>\n",
       "      <td>0.009048</td>\n",
       "      <td>0.013251</td>\n",
       "      <td>0.011520</td>\n",
       "      <td>0.016953</td>\n",
       "      <td>27</td>\n",
       "      <td>19</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.166679</td>\n",
       "      <td>0.856517</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>522922</td>\n",
       "      <td>3972</td>\n",
       "      <td>3925</td>\n",
       "      <td>3980</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002461</td>\n",
       "      <td>0.021451</td>\n",
       "      <td>0.021304</td>\n",
       "      <td>0.018869</td>\n",
       "      <td>0.009405</td>\n",
       "      <td>0.009381</td>\n",
       "      <td>0.007250</td>\n",
       "      <td>135</td>\n",
       "      <td>107</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.178789</td>\n",
       "      <td>0.821990</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>294752</td>\n",
       "      <td>2361</td>\n",
       "      <td>2362</td>\n",
       "      <td>2394</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001882</td>\n",
       "      <td>0.015978</td>\n",
       "      <td>0.019741</td>\n",
       "      <td>0.005820</td>\n",
       "      <td>0.007666</td>\n",
       "      <td>0.009325</td>\n",
       "      <td>0.003350</td>\n",
       "      <td>65</td>\n",
       "      <td>64</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>0.550811</td>\n",
       "      <td>0.881356</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>522918</td>\n",
       "      <td>1007</td>\n",
       "      <td>1009</td>\n",
       "      <td>1021</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.009699</td>\n",
       "      <td>0.009985</td>\n",
       "      <td>0.002079</td>\n",
       "      <td>0.004475</td>\n",
       "      <td>0.004838</td>\n",
       "      <td>0.000719</td>\n",
       "      <td>16</td>\n",
       "      <td>14</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>0.619598</td>\n",
       "      <td>0.764423</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "      <td>DLA3</td>\n",
       "      <td>1015</td>\n",
       "      <td>1640</td>\n",
       "      <td>1242</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000460</td>\n",
       "      <td>0.006302</td>\n",
       "      <td>0.002835</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004440</td>\n",
       "      <td>0.002568</td>\n",
       "      <td>0.000296</td>\n",
       "      <td>16</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0.938100</td>\n",
       "      <td>0.872832</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>DRG1</td>\n",
       "      <td>762</td>\n",
       "      <td>853</td>\n",
       "      <td>946</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.009381</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002314</td>\n",
       "      <td>0.002235</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000776</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>0.273290</td>\n",
       "      <td>0.785124</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>196073</td>\n",
       "      <td>1158</td>\n",
       "      <td>1161</td>\n",
       "      <td>1172</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000931</td>\n",
       "      <td>0.014366</td>\n",
       "      <td>0.016269</td>\n",
       "      <td>0.001552</td>\n",
       "      <td>0.005096</td>\n",
       "      <td>0.005397</td>\n",
       "      <td>0.000939</td>\n",
       "      <td>21</td>\n",
       "      <td>18</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>0.442575</td>\n",
       "      <td>0.804124</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>HDH1</td>\n",
       "      <td>625</td>\n",
       "      <td>714</td>\n",
       "      <td>827</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000714</td>\n",
       "      <td>0.016119</td>\n",
       "      <td>0.015761</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006269</td>\n",
       "      <td>0.006798</td>\n",
       "      <td>0.000888</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>0.739989</td>\n",
       "      <td>0.803279</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>NMDA1</td>\n",
       "      <td>412</td>\n",
       "      <td>430</td>\n",
       "      <td>551</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001036</td>\n",
       "      <td>0.006430</td>\n",
       "      <td>0.003801</td>\n",
       "      <td>0.003118</td>\n",
       "      <td>0.004780</td>\n",
       "      <td>0.001637</td>\n",
       "      <td>0.001992</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>0.860084</td>\n",
       "      <td>0.820225</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>GCSH</td>\n",
       "      <td>377</td>\n",
       "      <td>456</td>\n",
       "      <td>404</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001388</td>\n",
       "      <td>0.012623</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003745</td>\n",
       "      <td>0.000773</td>\n",
       "      <td>0.000908</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>0.089001</td>\n",
       "      <td>0.781316</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>PSF2</td>\n",
       "      <td>1010</td>\n",
       "      <td>1217</td>\n",
       "      <td>1864</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000300</td>\n",
       "      <td>0.027311</td>\n",
       "      <td>0.020161</td>\n",
       "      <td>0.021542</td>\n",
       "      <td>0.007986</td>\n",
       "      <td>0.006557</td>\n",
       "      <td>0.006591</td>\n",
       "      <td>29</td>\n",
       "      <td>23</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>0.112287</td>\n",
       "      <td>0.874396</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>344092</td>\n",
       "      <td>2542</td>\n",
       "      <td>2531</td>\n",
       "      <td>2448</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001631</td>\n",
       "      <td>0.012937</td>\n",
       "      <td>0.016012</td>\n",
       "      <td>0.006789</td>\n",
       "      <td>0.005411</td>\n",
       "      <td>0.006713</td>\n",
       "      <td>0.003129</td>\n",
       "      <td>49</td>\n",
       "      <td>49</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>0.114209</td>\n",
       "      <td>0.924000</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>196063</td>\n",
       "      <td>1149</td>\n",
       "      <td>1149</td>\n",
       "      <td>1161</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014878</td>\n",
       "      <td>0.018381</td>\n",
       "      <td>0.007359</td>\n",
       "      <td>0.005854</td>\n",
       "      <td>0.007232</td>\n",
       "      <td>0.003764</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>0.361543</td>\n",
       "      <td>0.708580</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>522914</td>\n",
       "      <td>2679</td>\n",
       "      <td>2701</td>\n",
       "      <td>2614</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003878</td>\n",
       "      <td>0.007664</td>\n",
       "      <td>0.003166</td>\n",
       "      <td>0.006246</td>\n",
       "      <td>0.005551</td>\n",
       "      <td>0.003242</td>\n",
       "      <td>0.004998</td>\n",
       "      <td>53</td>\n",
       "      <td>25</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>0.201471</td>\n",
       "      <td>0.724138</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>182389</td>\n",
       "      <td>399</td>\n",
       "      <td>400</td>\n",
       "      <td>401</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004181</td>\n",
       "      <td>0.009866</td>\n",
       "      <td>0.012189</td>\n",
       "      <td>0.008458</td>\n",
       "      <td>0.008354</td>\n",
       "      <td>0.009461</td>\n",
       "      <td>0.009824</td>\n",
       "      <td>12</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>0.295581</td>\n",
       "      <td>0.774590</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>CGL70</td>\n",
       "      <td>901</td>\n",
       "      <td>905</td>\n",
       "      <td>1347</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007198</td>\n",
       "      <td>0.008837</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004378</td>\n",
       "      <td>0.005006</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>0.171020</td>\n",
       "      <td>0.838942</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>HRGP1</td>\n",
       "      <td>1304</td>\n",
       "      <td>1342</td>\n",
       "      <td>1830</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000311</td>\n",
       "      <td>0.008759</td>\n",
       "      <td>0.010640</td>\n",
       "      <td>0.001795</td>\n",
       "      <td>0.002824</td>\n",
       "      <td>0.003133</td>\n",
       "      <td>0.000802</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>0.380677</td>\n",
       "      <td>0.933333</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>PTC1</td>\n",
       "      <td>1566</td>\n",
       "      <td>1725</td>\n",
       "      <td>1864</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.017480</td>\n",
       "      <td>0.013596</td>\n",
       "      <td>0.013141</td>\n",
       "      <td>0.005378</td>\n",
       "      <td>0.003851</td>\n",
       "      <td>0.003713</td>\n",
       "      <td>30</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>0.265610</td>\n",
       "      <td>0.855457</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>FUM1</td>\n",
       "      <td>1734</td>\n",
       "      <td>1740</td>\n",
       "      <td>1741</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001306</td>\n",
       "      <td>0.014348</td>\n",
       "      <td>0.013556</td>\n",
       "      <td>0.010886</td>\n",
       "      <td>0.005819</td>\n",
       "      <td>0.005007</td>\n",
       "      <td>0.004179</td>\n",
       "      <td>36</td>\n",
       "      <td>25</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>0.355551</td>\n",
       "      <td>0.820874</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>294742</td>\n",
       "      <td>2891</td>\n",
       "      <td>2869</td>\n",
       "      <td>2904</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001200</td>\n",
       "      <td>0.016782</td>\n",
       "      <td>0.017983</td>\n",
       "      <td>0.005325</td>\n",
       "      <td>0.007047</td>\n",
       "      <td>0.007477</td>\n",
       "      <td>0.002766</td>\n",
       "      <td>73</td>\n",
       "      <td>62</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>0.675902</td>\n",
       "      <td>0.797468</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>SPL2</td>\n",
       "      <td>566</td>\n",
       "      <td>572</td>\n",
       "      <td>864</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.019732</td>\n",
       "      <td>0.012189</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006424</td>\n",
       "      <td>0.004274</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>13</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>0.718422</td>\n",
       "      <td>0.791908</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>TOC34</td>\n",
       "      <td>543</td>\n",
       "      <td>827</td>\n",
       "      <td>689</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013005</td>\n",
       "      <td>0.002281</td>\n",
       "      <td>0.002875</td>\n",
       "      <td>0.006186</td>\n",
       "      <td>0.002967</td>\n",
       "      <td>0.000533</td>\n",
       "      <td>12</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>0.939833</td>\n",
       "      <td>0.755556</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>PR46b</td>\n",
       "      <td>241</td>\n",
       "      <td>567</td>\n",
       "      <td>366</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.011220</td>\n",
       "      <td>0.006202</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003518</td>\n",
       "      <td>0.003090</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>0.261433</td>\n",
       "      <td>0.907975</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>FBX9</td>\n",
       "      <td>881</td>\n",
       "      <td>883</td>\n",
       "      <td>892</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000634</td>\n",
       "      <td>0.010532</td>\n",
       "      <td>0.013012</td>\n",
       "      <td>0.002257</td>\n",
       "      <td>0.005103</td>\n",
       "      <td>0.005517</td>\n",
       "      <td>0.002051</td>\n",
       "      <td>16</td>\n",
       "      <td>14</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>0.291310</td>\n",
       "      <td>0.804979</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>ALB3</td>\n",
       "      <td>857</td>\n",
       "      <td>1088</td>\n",
       "      <td>876</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001309</td>\n",
       "      <td>0.018070</td>\n",
       "      <td>0.017823</td>\n",
       "      <td>0.009532</td>\n",
       "      <td>0.006843</td>\n",
       "      <td>0.007006</td>\n",
       "      <td>0.003742</td>\n",
       "      <td>21</td>\n",
       "      <td>22</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>0.388553</td>\n",
       "      <td>0.909532</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>UTP1</td>\n",
       "      <td>2585</td>\n",
       "      <td>2600</td>\n",
       "      <td>2917</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000599</td>\n",
       "      <td>0.009742</td>\n",
       "      <td>0.010731</td>\n",
       "      <td>0.001193</td>\n",
       "      <td>0.005002</td>\n",
       "      <td>0.005092</td>\n",
       "      <td>0.000755</td>\n",
       "      <td>46</td>\n",
       "      <td>38</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>0.508345</td>\n",
       "      <td>0.709239</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>522919</td>\n",
       "      <td>1501</td>\n",
       "      <td>1504</td>\n",
       "      <td>1530</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001091</td>\n",
       "      <td>0.016327</td>\n",
       "      <td>0.019211</td>\n",
       "      <td>0.001000</td>\n",
       "      <td>0.006883</td>\n",
       "      <td>0.007814</td>\n",
       "      <td>0.000959</td>\n",
       "      <td>37</td>\n",
       "      <td>34</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>0.326124</td>\n",
       "      <td>0.627660</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>MT0796</td>\n",
       "      <td>211</td>\n",
       "      <td>425</td>\n",
       "      <td>214</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007612</td>\n",
       "      <td>0.006812</td>\n",
       "      <td>0.011403</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006623</td>\n",
       "      <td>0.008918</td>\n",
       "      <td>0.005087</td>\n",
       "      <td>5</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>0.662986</td>\n",
       "      <td>0.873134</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>PDK1</td>\n",
       "      <td>785</td>\n",
       "      <td>886</td>\n",
       "      <td>1370</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013176</td>\n",
       "      <td>0.012550</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007108</td>\n",
       "      <td>0.007803</td>\n",
       "      <td>0.000268</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>0.097500</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>182392</td>\n",
       "      <td>149</td>\n",
       "      <td>334</td>\n",
       "      <td>151</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008415</td>\n",
       "      <td>0.010100</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003790</td>\n",
       "      <td>0.006238</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>0.075943</td>\n",
       "      <td>0.897638</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>182390</td>\n",
       "      <td>500</td>\n",
       "      <td>634</td>\n",
       "      <td>509</td>\n",
       "      <td>...</td>\n",
       "      <td>0.005525</td>\n",
       "      <td>0.020230</td>\n",
       "      <td>0.014027</td>\n",
       "      <td>0.026016</td>\n",
       "      <td>0.011004</td>\n",
       "      <td>0.009231</td>\n",
       "      <td>0.013911</td>\n",
       "      <td>20</td>\n",
       "      <td>17</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>0.097500</td>\n",
       "      <td>0.697479</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>97782</td>\n",
       "      <td>588</td>\n",
       "      <td>588</td>\n",
       "      <td>598</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.016830</td>\n",
       "      <td>0.020793</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004785</td>\n",
       "      <td>0.005911</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>0.619099</td>\n",
       "      <td>0.766513</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "      <td>522872</td>\n",
       "      <td>823</td>\n",
       "      <td>996</td>\n",
       "      <td>2653</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013071</td>\n",
       "      <td>0.009262</td>\n",
       "      <td>0.000594</td>\n",
       "      <td>0.006456</td>\n",
       "      <td>0.005932</td>\n",
       "      <td>0.000277</td>\n",
       "      <td>19</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>0.701355</td>\n",
       "      <td>0.754491</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>UBCH1</td>\n",
       "      <td>788</td>\n",
       "      <td>810</td>\n",
       "      <td>983</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012350</td>\n",
       "      <td>0.005050</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005345</td>\n",
       "      <td>0.002599</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>15</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>0.054931</td>\n",
       "      <td>0.790890</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>SPP3</td>\n",
       "      <td>1242</td>\n",
       "      <td>1579</td>\n",
       "      <td>1807</td>\n",
       "      <td>...</td>\n",
       "      <td>0.005885</td>\n",
       "      <td>0.026379</td>\n",
       "      <td>0.024754</td>\n",
       "      <td>0.024409</td>\n",
       "      <td>0.012975</td>\n",
       "      <td>0.013148</td>\n",
       "      <td>0.012586</td>\n",
       "      <td>59</td>\n",
       "      <td>61</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>0.046821</td>\n",
       "      <td>0.966667</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>SPS1</td>\n",
       "      <td>555</td>\n",
       "      <td>557</td>\n",
       "      <td>707</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015059</td>\n",
       "      <td>0.018411</td>\n",
       "      <td>0.012368</td>\n",
       "      <td>0.004066</td>\n",
       "      <td>0.005005</td>\n",
       "      <td>0.004117</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>0.112856</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>182394</td>\n",
       "      <td>721</td>\n",
       "      <td>856</td>\n",
       "      <td>728</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000786</td>\n",
       "      <td>0.005365</td>\n",
       "      <td>0.005641</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003140</td>\n",
       "      <td>0.003678</td>\n",
       "      <td>0.000505</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>0.121550</td>\n",
       "      <td>0.879781</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>THI10</td>\n",
       "      <td>812</td>\n",
       "      <td>812</td>\n",
       "      <td>820</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002782</td>\n",
       "      <td>0.017198</td>\n",
       "      <td>0.021248</td>\n",
       "      <td>0.014074</td>\n",
       "      <td>0.007213</td>\n",
       "      <td>0.008911</td>\n",
       "      <td>0.005742</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>0.881442</td>\n",
       "      <td>0.781977</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>23</td>\n",
       "      <td>MADS2</td>\n",
       "      <td>634</td>\n",
       "      <td>1202</td>\n",
       "      <td>1000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.022295</td>\n",
       "      <td>0.003774</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012515</td>\n",
       "      <td>0.002917</td>\n",
       "      <td>0.000368</td>\n",
       "      <td>29</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>0.131482</td>\n",
       "      <td>0.816092</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>CGLD28</td>\n",
       "      <td>640</td>\n",
       "      <td>631</td>\n",
       "      <td>645</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001748</td>\n",
       "      <td>0.031977</td>\n",
       "      <td>0.038100</td>\n",
       "      <td>0.024235</td>\n",
       "      <td>0.011170</td>\n",
       "      <td>0.013471</td>\n",
       "      <td>0.007817</td>\n",
       "      <td>26</td>\n",
       "      <td>25</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>0.764803</td>\n",
       "      <td>0.776699</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>LEU1S</td>\n",
       "      <td>511</td>\n",
       "      <td>561</td>\n",
       "      <td>561</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006812</td>\n",
       "      <td>0.011046</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003320</td>\n",
       "      <td>0.003123</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>0.709325</td>\n",
       "      <td>0.642308</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>13</td>\n",
       "      <td>522875</td>\n",
       "      <td>956</td>\n",
       "      <td>969</td>\n",
       "      <td>1251</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012561</td>\n",
       "      <td>0.006897</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006150</td>\n",
       "      <td>0.002894</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>21</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>0.151417</td>\n",
       "      <td>0.724101</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>SAD1</td>\n",
       "      <td>11202</td>\n",
       "      <td>10858</td>\n",
       "      <td>11172</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003761</td>\n",
       "      <td>0.020827</td>\n",
       "      <td>0.024001</td>\n",
       "      <td>0.012588</td>\n",
       "      <td>0.010932</td>\n",
       "      <td>0.012256</td>\n",
       "      <td>0.006820</td>\n",
       "      <td>445</td>\n",
       "      <td>390</td>\n",
       "      <td>211</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>50 rows  48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Fst       GC4  alleles_ALL  alleles_MTMINUS  alleles_MTPLUS  fixed  \\\n",
       "0   0.034259  0.866279           19               10               9      0   \n",
       "1   0.392467  0.809091           19               10               9      2   \n",
       "2   0.466787  0.976190           19               10               9      0   \n",
       "3   0.131916  0.854801           19               10               9      2   \n",
       "4   0.092847  0.798701           19               10               9      0   \n",
       "5   0.660817  0.779241           19               10               9     64   \n",
       "6   0.583469  0.517949           19               10               9      5   \n",
       "7   0.315672  0.718750           19               10               9      0   \n",
       "8   0.166679  0.856517           19               10               9      0   \n",
       "9   0.178789  0.821990           19               10               9      0   \n",
       "10  0.550811  0.881356           19               10               9      0   \n",
       "11  0.619598  0.764423           19               10               9      8   \n",
       "12  0.938100  0.872832           19               10               9      5   \n",
       "13  0.273290  0.785124           19               10               9      0   \n",
       "14  0.442575  0.804124           19               10               9      1   \n",
       "15  0.739989  0.803279           19               10               9      5   \n",
       "16  0.860084  0.820225           19               10               9      3   \n",
       "17  0.089001  0.781316           19               10               9      0   \n",
       "18  0.112287  0.874396           19               10               9      0   \n",
       "19  0.114209  0.924000           19               10               9      0   \n",
       "20  0.361543  0.708580           19               10               9      0   \n",
       "21  0.201471  0.724138           19               10               9      0   \n",
       "22  0.295581  0.774590           19               10               9      1   \n",
       "23  0.171020  0.838942           19               10               9      0   \n",
       "24  0.380677  0.933333           19               10               9     10   \n",
       "25  0.265610  0.855457           19               10               9      0   \n",
       "26  0.355551  0.820874           19               10               9      0   \n",
       "27  0.675902  0.797468           19               10               9      6   \n",
       "28  0.718422  0.791908           19               10               9      7   \n",
       "29  0.939833  0.755556           19               10               9      2   \n",
       "30  0.261433  0.907975           19               10               9      0   \n",
       "31  0.291310  0.804979           19               10               9      2   \n",
       "32  0.388553  0.909532           19               10               9      6   \n",
       "33  0.508345  0.709239           19               10               9      0   \n",
       "34  0.326124  0.627660           19               10               9      0   \n",
       "35  0.662986  0.873134           19               10               9      4   \n",
       "36  0.097500  0.714286           19               10               9      0   \n",
       "37  0.075943  0.897638           19               10               9      0   \n",
       "38  0.097500  0.697479           19               10               9      0   \n",
       "39  0.619099  0.766513           19               10               9      8   \n",
       "40  0.701355  0.754491           19               10               9      9   \n",
       "41  0.054931  0.790890           19               10               9      1   \n",
       "42  0.046821  0.966667           19               10               9      0   \n",
       "43  0.112856  0.750000           19               10               9      0   \n",
       "44  0.121550  0.879781           19               10               9      0   \n",
       "45  0.881442  0.781977           19               10               9     23   \n",
       "46  0.131482  0.816092           19               10               9      0   \n",
       "47  0.764803  0.776699           19               10               9      3   \n",
       "48  0.709325  0.642308           19               10               9     13   \n",
       "49  0.151417  0.724101           19               10               9      0   \n",
       "\n",
       "      gene  invariant_ALL  invariant_MTMINUS  invariant_MTPLUS      ...        \\\n",
       "0   EIF5Bb            929               1080               991      ...         \n",
       "1   MT0829           1066               1075              1088      ...         \n",
       "2   MT0828            147                 77               169      ...         \n",
       "3     NIC7           1443               1716              1654      ...         \n",
       "4     RFC4            641                724               651      ...         \n",
       "5     PKY1           5275               6158              7876      ...         \n",
       "6     LPS1            654               2257              1666      ...         \n",
       "7   522915            556                564               559      ...         \n",
       "8   522922           3972               3925              3980      ...         \n",
       "9   294752           2361               2362              2394      ...         \n",
       "10  522918           1007               1009              1021      ...         \n",
       "11    DLA3           1015               1640              1242      ...         \n",
       "12    DRG1            762                853               946      ...         \n",
       "13  196073           1158               1161              1172      ...         \n",
       "14    HDH1            625                714               827      ...         \n",
       "15   NMDA1            412                430               551      ...         \n",
       "16    GCSH            377                456               404      ...         \n",
       "17    PSF2           1010               1217              1864      ...         \n",
       "18  344092           2542               2531              2448      ...         \n",
       "19  196063           1149               1149              1161      ...         \n",
       "20  522914           2679               2701              2614      ...         \n",
       "21  182389            399                400               401      ...         \n",
       "22   CGL70            901                905              1347      ...         \n",
       "23   HRGP1           1304               1342              1830      ...         \n",
       "24    PTC1           1566               1725              1864      ...         \n",
       "25    FUM1           1734               1740              1741      ...         \n",
       "26  294742           2891               2869              2904      ...         \n",
       "27    SPL2            566                572               864      ...         \n",
       "28   TOC34            543                827               689      ...         \n",
       "29   PR46b            241                567               366      ...         \n",
       "30    FBX9            881                883               892      ...         \n",
       "31    ALB3            857               1088               876      ...         \n",
       "32    UTP1           2585               2600              2917      ...         \n",
       "33  522919           1501               1504              1530      ...         \n",
       "34  MT0796            211                425               214      ...         \n",
       "35    PDK1            785                886              1370      ...         \n",
       "36  182392            149                334               151      ...         \n",
       "37  182390            500                634               509      ...         \n",
       "38   97782            588                588               598      ...         \n",
       "39  522872            823                996              2653      ...         \n",
       "40   UBCH1            788                810               983      ...         \n",
       "41    SPP3           1242               1579              1807      ...         \n",
       "42    SPS1            555                557               707      ...         \n",
       "43  182394            721                856               728      ...         \n",
       "44   THI10            812                812               820      ...         \n",
       "45   MADS2            634               1202              1000      ...         \n",
       "46  CGLD28            640                631               645      ...         \n",
       "47   LEU1S            511                561               561      ...         \n",
       "48  522875            956                969              1251      ...         \n",
       "49    SAD1          11202              10858             11172      ...         \n",
       "\n",
       "   theta_w_0fold_MTPLUS theta_w_4fold_ALL  theta_w_4fold_MTMINUS  \\\n",
       "0              0.000000          0.010146               0.008518   \n",
       "1              0.002562          0.018290               0.019369   \n",
       "2              0.010316          0.016830               0.000000   \n",
       "3              0.000686          0.020373               0.016967   \n",
       "4              0.000918          0.014619               0.020658   \n",
       "5              0.000722          0.010128               0.005515   \n",
       "6              0.000338          0.017882               0.010941   \n",
       "7              0.018642          0.007036               0.008692   \n",
       "8              0.002461          0.021451               0.021304   \n",
       "9              0.001882          0.015978               0.019741   \n",
       "10             0.000000          0.009699               0.009985   \n",
       "11             0.000460          0.006302               0.002835   \n",
       "12             0.000000          0.009381               0.000000   \n",
       "13             0.000931          0.014366               0.016269   \n",
       "14             0.000714          0.016119               0.015761   \n",
       "15             0.001036          0.006430               0.003801   \n",
       "16             0.001388          0.012623               0.000000   \n",
       "17             0.000300          0.027311               0.020161   \n",
       "18             0.001631          0.012937               0.016012   \n",
       "19             0.000000          0.014878               0.018381   \n",
       "20             0.003878          0.007664               0.003166   \n",
       "21             0.004181          0.009866               0.012189   \n",
       "22             0.000000          0.007198               0.008837   \n",
       "23             0.000311          0.008759               0.010640   \n",
       "24             0.000000          0.017480               0.013596   \n",
       "25             0.001306          0.014348               0.013556   \n",
       "26             0.001200          0.016782               0.017983   \n",
       "27             0.000000          0.019732               0.012189   \n",
       "28             0.000000          0.013005               0.002281   \n",
       "29             0.000000          0.011220               0.006202   \n",
       "30             0.000634          0.010532               0.013012   \n",
       "31             0.001309          0.018070               0.017823   \n",
       "32             0.000599          0.009742               0.010731   \n",
       "33             0.001091          0.016327               0.019211   \n",
       "34             0.007612          0.006812               0.011403   \n",
       "35             0.000000          0.013176               0.012550   \n",
       "36             0.000000          0.008415               0.010100   \n",
       "37             0.005525          0.020230               0.014027   \n",
       "38             0.000000          0.016830               0.020793   \n",
       "39             0.000000          0.013071               0.009262   \n",
       "40             0.000000          0.012350               0.005050   \n",
       "41             0.005885          0.026379               0.024754   \n",
       "42             0.000000          0.015059               0.018411   \n",
       "43             0.000786          0.005365               0.005641   \n",
       "44             0.002782          0.017198               0.021248   \n",
       "45             0.000000          0.022295               0.003774   \n",
       "46             0.001748          0.031977               0.038100   \n",
       "47             0.000000          0.006812               0.011046   \n",
       "48             0.000000          0.012561               0.006897   \n",
       "49             0.003761          0.020827               0.024001   \n",
       "\n",
       "    theta_w_4fold_MTPLUS theta_w_total_ALL theta_w_total_MTMINUS  \\\n",
       "0               0.010012          0.006030              0.006111   \n",
       "1               0.005017          0.007323              0.006139   \n",
       "2               0.017521          0.013005              0.000000   \n",
       "3               0.022688          0.007341              0.006670   \n",
       "4               0.000000          0.004395              0.005763   \n",
       "5               0.001018          0.006882              0.003411   \n",
       "6               0.001256          0.008490              0.005853   \n",
       "7               0.009048          0.013251              0.011520   \n",
       "8               0.018869          0.009405              0.009381   \n",
       "9               0.005820          0.007666              0.009325   \n",
       "10              0.002079          0.004475              0.004838   \n",
       "11              0.000000          0.004440              0.002568   \n",
       "12              0.002314          0.002235              0.000000   \n",
       "13              0.001552          0.005096              0.005397   \n",
       "14              0.000000          0.006269              0.006798   \n",
       "15              0.003118          0.004780              0.001637   \n",
       "16              0.000000          0.003745              0.000773   \n",
       "17              0.021542          0.007986              0.006557   \n",
       "18              0.006789          0.005411              0.006713   \n",
       "19              0.007359          0.005854              0.007232   \n",
       "20              0.006246          0.005551              0.003242   \n",
       "21              0.008458          0.008354              0.009461   \n",
       "22              0.000000          0.004378              0.005006   \n",
       "23              0.001795          0.002824              0.003133   \n",
       "24              0.013141          0.005378              0.003851   \n",
       "25              0.010886          0.005819              0.005007   \n",
       "26              0.005325          0.007047              0.007477   \n",
       "27              0.000000          0.006424              0.004274   \n",
       "28              0.002875          0.006186              0.002967   \n",
       "29              0.000000          0.003518              0.003090   \n",
       "30              0.002257          0.005103              0.005517   \n",
       "31              0.009532          0.006843              0.007006   \n",
       "32              0.001193          0.005002              0.005092   \n",
       "33              0.001000          0.006883              0.007814   \n",
       "34              0.000000          0.006623              0.008918   \n",
       "35              0.000000          0.007108              0.007803   \n",
       "36              0.000000          0.003790              0.006238   \n",
       "37              0.026016          0.011004              0.009231   \n",
       "38              0.000000          0.004785              0.005911   \n",
       "39              0.000594          0.006456              0.005932   \n",
       "40              0.000000          0.005345              0.002599   \n",
       "41              0.024409          0.012975              0.013148   \n",
       "42              0.012368          0.004066              0.005005   \n",
       "43              0.000000          0.003140              0.003678   \n",
       "44              0.014074          0.007213              0.008911   \n",
       "45              0.000000          0.012515              0.002917   \n",
       "46              0.024235          0.011170              0.013471   \n",
       "47              0.000000          0.003320              0.003123   \n",
       "48              0.000000          0.006150              0.002894   \n",
       "49              0.012588          0.010932              0.012256   \n",
       "\n",
       "   theta_w_total_MTPLUS variant_ALL variant_MTMINUS variant_MTPLUS  \n",
       "0              0.005846          20              19             16  \n",
       "1              0.003351          28              19             10  \n",
       "2              0.010573           7               0              5  \n",
       "3              0.008688          38              33             40  \n",
       "4              0.000564          10              12              1  \n",
       "5              0.000839         130              60             18  \n",
       "6              0.000441          20              38              2  \n",
       "7              0.016953          27              19             27  \n",
       "8              0.007250         135             107             80  \n",
       "9              0.003350          65              64             22  \n",
       "10             0.000719          16              14              2  \n",
       "11             0.000296          16              12              1  \n",
       "12             0.000776           6               0              2  \n",
       "13             0.000939          21              18              3  \n",
       "14             0.000888          14              14              2  \n",
       "15             0.001992           7               2              3  \n",
       "16             0.000908           5               1              1  \n",
       "17             0.006591          29              23             34  \n",
       "18             0.003129          49              49             21  \n",
       "19             0.003764          24              24             12  \n",
       "20             0.004998          53              25             36  \n",
       "21             0.009824          12              11             11  \n",
       "22             0.000000          14              13              0  \n",
       "23             0.000802          13              12              4  \n",
       "24             0.003713          30              19             19  \n",
       "25             0.004179          36              25             20  \n",
       "26             0.002766          73              62             22  \n",
       "27             0.000000          13               7              0  \n",
       "28             0.000533          12               7              1  \n",
       "29             0.000000           3               5              0  \n",
       "30             0.002051          16              14              5  \n",
       "31             0.003742          21              22              9  \n",
       "32             0.000755          46              38              6  \n",
       "33             0.000959          37              34              4  \n",
       "34             0.005087           5              11              3  \n",
       "35             0.000268          20              20              1  \n",
       "36             0.000000           2               6              0  \n",
       "37             0.013911          20              17             20  \n",
       "38             0.000000          10              10              0  \n",
       "39             0.000277          19              17              2  \n",
       "40             0.000000          15               6              0  \n",
       "41             0.012586          59              61             64  \n",
       "42             0.004117           8               8              8  \n",
       "43             0.000505           8               9              1  \n",
       "44             0.005742          21              21             13  \n",
       "45             0.000368          29              10              1  \n",
       "46             0.007817          26              25             14  \n",
       "47             0.000000           6               5              0  \n",
       "48             0.000000          21               8              0  \n",
       "49             0.006820         445             390            211  \n",
       "\n",
       "[50 rows x 48 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "shared_results.head(n=50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# shared_results.to_csv('polymorphism_all.shared.csv', index = False)\n",
    "shared_results.to_csv('polymorphism_all.20190225.shared.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# mtLimited transcripts\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../VCF2FASTA/mtLimited/MTP0428.26893319.quebec_mtPlus.fasta\n",
      "MTP0428 mtPlus\n",
      "../VCF2FASTA/mtLimited/MTA3.26894517.quebec_mtPlus.fasta\n",
      "MTA3 mtPlus\n",
      "../VCF2FASTA/mtLimited/MTA4.26893740.quebec_mtPlus.fasta\n",
      "MTA4 mtPlus\n",
      "../VCF2FASTA/mtLimited/MTA1.26893117.quebec_mtPlus.fasta\n",
      "MTA1 mtPlus\n",
      "../VCF2FASTA/mtLimited/INT1a.26894683.quebec_mtPlus.fasta\n",
      "INT1a mtPlus\n",
      "../VCF2FASTA/mtLimited/294708.26893164.quebec_mtPlus.fasta\n",
      "294708 mtPlus\n",
      "../VCF2FASTA/mtLimited/EZY2c.26894249.quebec_mtPlus.fasta\n",
      "EZY2c mtPlus\n",
      "../VCF2FASTA/mtLimited/MID.ADF43190.1.quebec_mtMinus.fasta\n",
      "MID mtMinus\n",
      "../VCF2FASTA/mtLimited/MTA5.26894560.quebec_mtPlus.fasta\n",
      "MTA5 mtPlus\n",
      "../VCF2FASTA/mtLimited/FUS1.26894229.quebec_mtPlus.fasta\n",
      "FUS1 mtPlus\n",
      "../VCF2FASTA/mtLimited/MTD1.ADF43193.1.quebec_mtMinus.fasta\n",
      "MTD1 mtMinus\n",
      "../VCF2FASTA/mtLimited/INT1b.26893202.quebec_mtPlus.fasta\n",
      "INT1b mtPlus\n",
      "../VCF2FASTA/mtLimited/EZY2d.26893701.quebec_mtPlus.fasta\n",
      "EZY2d mtPlus\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GC4</th>\n",
       "      <th>alleles_ALL</th>\n",
       "      <th>gene</th>\n",
       "      <th>invariant_ALL</th>\n",
       "      <th>mtMinus_coords</th>\n",
       "      <th>mtPlus_coords</th>\n",
       "      <th>sfs_0fold_ALL</th>\n",
       "      <th>sfs_4fold_ALL</th>\n",
       "      <th>sfs_t_ALL</th>\n",
       "      <th>sites_ALL</th>\n",
       "      <th>theta_pi_0fold_ALL</th>\n",
       "      <th>theta_pi_4fold_ALL</th>\n",
       "      <th>theta_pi_total_ALL</th>\n",
       "      <th>theta_w_0fold_ALL</th>\n",
       "      <th>theta_w_4fold_ALL</th>\n",
       "      <th>theta_w_total_ALL</th>\n",
       "      <th>variant_ALL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.717836</td>\n",
       "      <td>9</td>\n",
       "      <td>MTP0428</td>\n",
       "      <td>3110</td>\n",
       "      <td>N/A</td>\n",
       "      <td>chromosome_6:425999-429941</td>\n",
       "      <td>[1941, 4, 3, 5, 0, 1, 5, 2, 5, 0]</td>\n",
       "      <td>[675, 2, 4, 1, 0, 0, 0, 1, 1, 0]</td>\n",
       "      <td>[3110, 8, 10, 7, 0, 1, 5, 4, 7, 0]</td>\n",
       "      <td>3152</td>\n",
       "      <td>0.004832</td>\n",
       "      <td>0.004548</td>\n",
       "      <td>0.004865</td>\n",
       "      <td>0.004679</td>\n",
       "      <td>0.004841</td>\n",
       "      <td>0.004903</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.756410</td>\n",
       "      <td>9</td>\n",
       "      <td>MTA3</td>\n",
       "      <td>380</td>\n",
       "      <td>N/A</td>\n",
       "      <td>chromosome_6:563911-564726</td>\n",
       "      <td>[244, 0, 0, 0, 0, 0, 0, 0, 0, 0]</td>\n",
       "      <td>[76, 1, 0, 0, 0, 0, 0, 0, 1, 0]</td>\n",
       "      <td>[380, 1, 0, 0, 0, 0, 0, 1, 1, 0]</td>\n",
       "      <td>383</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005698</td>\n",
       "      <td>0.002176</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.009434</td>\n",
       "      <td>0.002882</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.880531</td>\n",
       "      <td>9</td>\n",
       "      <td>MTA4</td>\n",
       "      <td>1063</td>\n",
       "      <td>N/A</td>\n",
       "      <td>chromosome_6:548791-552710</td>\n",
       "      <td>[673, 0, 0, 0, 0, 0, 0, 0, 0, 0]</td>\n",
       "      <td>[221, 0, 2, 0, 0, 0, 0, 1, 2, 0]</td>\n",
       "      <td>[1063, 3, 2, 0, 0, 0, 0, 2, 3, 0]</td>\n",
       "      <td>1073</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007129</td>\n",
       "      <td>0.002692</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008140</td>\n",
       "      <td>0.003429</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.790323</td>\n",
       "      <td>9</td>\n",
       "      <td>MTA1</td>\n",
       "      <td>358</td>\n",
       "      <td>N/A</td>\n",
       "      <td>chromosome_6:559864-560244</td>\n",
       "      <td>[221, 0, 0, 0, 0, 0, 0, 0, 0, 0]</td>\n",
       "      <td>[62, 0, 0, 0, 0, 0, 0, 0, 0, 0]</td>\n",
       "      <td>[358, 0, 0, 0, 0, 0, 0, 0, 0, 0]</td>\n",
       "      <td>358</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.849138</td>\n",
       "      <td>9</td>\n",
       "      <td>INT1a</td>\n",
       "      <td>1087</td>\n",
       "      <td>N/A</td>\n",
       "      <td>chromosome_6:740848-742579</td>\n",
       "      <td>[702, 0, 0, 0, 0, 0, 0, 0, 1, 0]</td>\n",
       "      <td>[232, 0, 0, 0, 0, 0, 0, 0, 0, 0]</td>\n",
       "      <td>[1087, 0, 0, 0, 0, 0, 0, 0, 1, 0]</td>\n",
       "      <td>1088</td>\n",
       "      <td>0.000316</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000204</td>\n",
       "      <td>0.000523</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000338</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        GC4  alleles_ALL     gene  invariant_ALL mtMinus_coords  \\\n",
       "0  0.717836            9  MTP0428           3110            N/A   \n",
       "1  0.756410            9     MTA3            380            N/A   \n",
       "2  0.880531            9     MTA4           1063            N/A   \n",
       "3  0.790323            9     MTA1            358            N/A   \n",
       "4  0.849138            9    INT1a           1087            N/A   \n",
       "\n",
       "                mtPlus_coords                      sfs_0fold_ALL  \\\n",
       "0  chromosome_6:425999-429941  [1941, 4, 3, 5, 0, 1, 5, 2, 5, 0]   \n",
       "1  chromosome_6:563911-564726   [244, 0, 0, 0, 0, 0, 0, 0, 0, 0]   \n",
       "2  chromosome_6:548791-552710   [673, 0, 0, 0, 0, 0, 0, 0, 0, 0]   \n",
       "3  chromosome_6:559864-560244   [221, 0, 0, 0, 0, 0, 0, 0, 0, 0]   \n",
       "4  chromosome_6:740848-742579   [702, 0, 0, 0, 0, 0, 0, 0, 1, 0]   \n",
       "\n",
       "                      sfs_4fold_ALL                           sfs_t_ALL  \\\n",
       "0  [675, 2, 4, 1, 0, 0, 0, 1, 1, 0]  [3110, 8, 10, 7, 0, 1, 5, 4, 7, 0]   \n",
       "1   [76, 1, 0, 0, 0, 0, 0, 0, 1, 0]    [380, 1, 0, 0, 0, 0, 0, 1, 1, 0]   \n",
       "2  [221, 0, 2, 0, 0, 0, 0, 1, 2, 0]   [1063, 3, 2, 0, 0, 0, 0, 2, 3, 0]   \n",
       "3   [62, 0, 0, 0, 0, 0, 0, 0, 0, 0]    [358, 0, 0, 0, 0, 0, 0, 0, 0, 0]   \n",
       "4  [232, 0, 0, 0, 0, 0, 0, 0, 0, 0]   [1087, 0, 0, 0, 0, 0, 0, 0, 1, 0]   \n",
       "\n",
       "   sites_ALL  theta_pi_0fold_ALL  theta_pi_4fold_ALL  theta_pi_total_ALL  \\\n",
       "0       3152            0.004832            0.004548            0.004865   \n",
       "1        383            0.000000            0.005698            0.002176   \n",
       "2       1073            0.000000            0.007129            0.002692   \n",
       "3        358            0.000000            0.000000            0.000000   \n",
       "4       1088            0.000316            0.000000            0.000204   \n",
       "\n",
       "   theta_w_0fold_ALL  theta_w_4fold_ALL  theta_w_total_ALL  variant_ALL  \n",
       "0           0.004679           0.004841           0.004903           42  \n",
       "1           0.000000           0.009434           0.002882            3  \n",
       "2           0.000000           0.008140           0.003429           10  \n",
       "3           0.000000           0.000000           0.000000            0  \n",
       "4           0.000523           0.000000           0.000338            1  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Target Genotype Lists\n",
    "mtPlus_Genotypes = \"CC2936 CC2937 CC3060 CC3064 CC3065 CC3068 CC3071 CC3076 CC3086\".split()\n",
    "mtMinus_Genotypes = \"CC2935 CC2938 CC3059 CC3061 CC3062 CC3063 CC3073 CC3075 CC3079 CC3084\".split()\n",
    "\n",
    "\n",
    "#### STATS TO COLLECT for each gene\n",
    "genes = []\n",
    "mtPlus_coords = []\n",
    "mtMinus_coords = []\n",
    "alleles_ALL = []\n",
    "sites_ALL = []\n",
    "invariant_ALL = []\n",
    "variant_ALL = []\n",
    "theta_pi_0fold_ALL = []\n",
    "theta_pi_4fold_ALL = []\n",
    "theta_pi_total_ALL = []\n",
    "theta_w_0fold_ALL = []\n",
    "theta_w_4fold_ALL = []\n",
    "theta_w_total_ALL = []\n",
    "sfs_4fold_ALL = []\n",
    "sfs_0fold_ALL = []\n",
    "sfs_t_ALL = []\n",
    "GC4 = []\n",
    "# Loop over all shared coding gametologs\n",
    "alignment_files = glob.glob('../VCF2FASTA/mtLimited/*.fasta')\n",
    "\n",
    "for a in alignment_files[:]:\n",
    "    print(a)\n",
    "    gene = a.split(\"/\")[-1].split(\".\")[0]\n",
    "    mt_allele = a.split(\"/\")[-1].split(\".\")[-2].split(\"_\")[-1]\n",
    "    print(gene, mt_allele)\n",
    "    genes.append(gene)\n",
    "    aln = AlignIO.read(open(a), 'fasta')\n",
    "    if mt_allele == \"mtPlus\":\n",
    "        mtPlus_sampleIDs  = [s.id for s in aln if s.id.split(\"|\")[0] in mtPlus_Genotypes]\n",
    "        mtPlus_coords.append(mtPlus_sampleIDs[0].split(\"|\")[-1]) ## based on this kind of name e.g. \"CC2935|mtMinus:105065-114642\"\n",
    "        mtMinus_coords.append('N/A')\n",
    "    else:\n",
    "        mtMinus_sampleIDs = [s.id for s in aln if s.id.split(\"|\")[0] in mtMinus_Genotypes]\n",
    "        mtMinus_coords.append(mtMinus_sampleIDs[0].split(\"|\")[-1])\n",
    "        mtPlus_coords.append(\"N/A\")\n",
    "    degen_map = degenerate(special_consensus(aln))\n",
    "    sfst,sfs0,sfs4 = alignment2sfs_w_degen(aln, 6, degen_map, samples=None, start=None, end=None)\n",
    "    ##### record ALL results\n",
    "    alleles_ALL.append(sfst.alleles)\n",
    "    sites_ALL.append(sfst.sites())\n",
    "    invariant_ALL.append(sfst.invariant())\n",
    "    variant_ALL.append(sfst.variant())\n",
    "    theta_pi_0fold_ALL.append(sfs0.theta_pi())\n",
    "    theta_pi_4fold_ALL.append(sfs4.theta_pi())\n",
    "    theta_pi_total_ALL.append(sfst.theta_pi())\n",
    "    theta_w_0fold_ALL.append(sfs0.theta_w())\n",
    "    theta_w_4fold_ALL.append(sfs4.theta_w())\n",
    "    theta_w_total_ALL.append(sfst.theta_w())\n",
    "    sfs_4fold_ALL.append(sfs4.sfs)\n",
    "    sfs_0fold_ALL.append(sfs0.sfs)\n",
    "    sfs_t_ALL.append(sfst.sfs)\n",
    "    #######\n",
    "    ## GC_4fold\n",
    "    GC4.append(len([1 for i,j in zip(special_consensus(aln),degen_map) if i in 'GCgc' and j =='4'])/\n",
    "               len([1 for i,j in zip(special_consensus(aln),degen_map) if i in 'AaTtGCgc' and j =='4']))\n",
    "\n",
    "\n",
    "    \n",
    "#########################################################\n",
    "#### DF the results #####################################\n",
    "#########################################################\n",
    "mtLimited_results = pd.DataFrame(\n",
    "    {'gene' : genes,\n",
    "     'mtPlus_coords' : mtPlus_coords,\n",
    "     'mtMinus_coords' : mtMinus_coords,\n",
    "     'alleles_ALL' : alleles_ALL, \n",
    "     'sites_ALL' : sites_ALL, \n",
    "     'invariant_ALL' : invariant_ALL, \n",
    "     'variant_ALL' : variant_ALL, \n",
    "     'theta_pi_0fold_ALL' : theta_pi_0fold_ALL, \n",
    "     'theta_pi_4fold_ALL' : theta_pi_4fold_ALL, \n",
    "     'theta_pi_total_ALL' : theta_pi_total_ALL, \n",
    "     'theta_w_0fold_ALL' : theta_w_0fold_ALL, \n",
    "     'theta_w_4fold_ALL' : theta_w_4fold_ALL, \n",
    "     'theta_w_total_ALL' : theta_w_total_ALL, \n",
    "     'sfs_4fold_ALL' : sfs_4fold_ALL, \n",
    "     'sfs_0fold_ALL' : sfs_0fold_ALL, \n",
    "     'sfs_t_ALL' : sfs_t_ALL,    \n",
    "     'GC4' : GC4\n",
    "    })  \n",
    "mtLimited_results.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "mtLimited_results.to_csv('polymorphism_all.mtLimited.csv', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ch6 and random transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Target Genotype Lists\n",
    "mtPlus_Genotypes = \"CC2936 CC2937 CC3060 CC3064 CC3065 CC3068 CC3071 CC3076 CC3086\".split()\n",
    "mtMinus_Genotypes = \"CC2935 CC2938 CC3059 CC3061 CC3062 CC3063 CC3073 CC3075 CC3079 CC3084\".split()\n",
    "\n",
    "\n",
    "#### STATS TO COLLECT for each gene\n",
    "genes = []\n",
    "mtPlus_coords = []\n",
    "mtMinus_coords = []\n",
    "alleles_ALL = []\n",
    "sites_ALL = []\n",
    "invariant_ALL = []\n",
    "variant_ALL = []\n",
    "theta_pi_0fold_ALL = []\n",
    "theta_pi_4fold_ALL = []\n",
    "theta_pi_total_ALL = []\n",
    "theta_w_0fold_ALL = []\n",
    "theta_w_4fold_ALL = []\n",
    "theta_w_total_ALL = []\n",
    "sfs_4fold_ALL = []\n",
    "sfs_0fold_ALL = []\n",
    "sfs_t_ALL = []\n",
    "alleles_MTPLUS = []\n",
    "sites_MTPLUS = []\n",
    "invariant_MTPLUS = []\n",
    "variant_MTPLUS = []\n",
    "theta_pi_0fold_MTPLUS = []\n",
    "theta_pi_4fold_MTPLUS = []\n",
    "theta_pi_total_MTPLUS = []\n",
    "theta_w_0fold_MTPLUS = []\n",
    "theta_w_4fold_MTPLUS = []\n",
    "theta_w_total_MTPLUS = []\n",
    "sfs_4fold_MTPLUS = []\n",
    "sfs_0fold_MTPLUS = []\n",
    "sfs_t_MTPLUS = []\n",
    "alleles_MTMINUS = []\n",
    "sites_MTMINUS = []\n",
    "invariant_MTMINUS = []\n",
    "variant_MTMINUS = []\n",
    "theta_pi_0fold_MTMINUS = []\n",
    "theta_pi_4fold_MTMINUS = []\n",
    "theta_pi_total_MTMINUS = []\n",
    "theta_w_0fold_MTMINUS = []\n",
    "theta_w_4fold_MTMINUS = []\n",
    "theta_w_total_MTMINUS = []\n",
    "sfs_4fold_MTMINUS = []\n",
    "sfs_0fold_MTMINUS = []\n",
    "sfs_t_MTMINUS = []\n",
    "private_MTPLUS = []\n",
    "private_MTMINUS = []\n",
    "shared = []\n",
    "fixed = []\n",
    "Fst = []\n",
    "GC4 = []\n",
    "\n",
    "# Loop over all shared coding gametologs\n",
    "ch6_alignment_files = glob.glob('../VCF2FASTA/chromosome_6/*fasta')\n",
    "random_alignment_files  =glob.glob('../VCF2FASTA/random_transcripts2/*fasta')\n",
    "\n",
    "bc = 0\n",
    "for a in ch6_alignment_files+random_alignment_files:\n",
    "    bc+=1\n",
    "    if bc %25 == 0:print(bc)\n",
    "    gene = a.split(\"/\")[-1].split(\".\")[0]\n",
    "    genes.append(gene)\n",
    "    aln = AlignIO.read(open(a), 'fasta')\n",
    "    mtPlus_sampleIDs  = [s.id for s in aln if s.id.split(\"|\")[0] in mtPlus_Genotypes]\n",
    "    mtMinus_sampleIDs = [s.id for s in aln if s.id.split(\"|\")[0] in mtMinus_Genotypes]\n",
    "    mtPlus_coords.append(mtPlus_sampleIDs[0].split(\"|\")[-1]) ## based on this kind of name e.g. \"CC2935|mtMinus:105065-114642\"\n",
    "    mtMinus_coords.append(mtMinus_sampleIDs[0].split(\"|\")[-1])\n",
    "    degen_map = degenerate(special_consensus(aln))\n",
    "    sfst,sfs0,sfs4 = alignment2sfs_w_degen(aln, 12, degen_map, samples=None, start=None, end=None)\n",
    "    ##### record ALL results\n",
    "    alleles_ALL.append(sfst.alleles)\n",
    "    sites_ALL.append(sfst.sites())\n",
    "    invariant_ALL.append(sfst.invariant())\n",
    "    variant_ALL.append(sfst.variant())\n",
    "    theta_pi_0fold_ALL.append(sfs0.theta_pi())\n",
    "    theta_pi_4fold_ALL.append(sfs4.theta_pi())\n",
    "    theta_pi_total_ALL.append(sfst.theta_pi())\n",
    "    theta_w_0fold_ALL.append(sfs0.theta_w())\n",
    "    theta_w_4fold_ALL.append(sfs4.theta_w())\n",
    "    theta_w_total_ALL.append(sfst.theta_w())\n",
    "    sfs_4fold_ALL.append(sfs4.sfs)\n",
    "    sfs_0fold_ALL.append(sfs0.sfs)\n",
    "    sfs_t_ALL.append(sfst.sfs)\n",
    "    ###################################\n",
    "    ## MTPLUS\n",
    "    sfst,sfs0,sfs4 = alignment2sfs_w_degen(aln, 8, degen_map, samples=mtPlus_sampleIDs, start=None, end=None)\n",
    "    ##### record ALL results\n",
    "    alleles_MTPLUS.append(sfst.alleles)\n",
    "    sites_MTPLUS.append(sfst.sites())\n",
    "    invariant_MTPLUS.append(sfst.invariant())\n",
    "    variant_MTPLUS.append(sfst.variant())\n",
    "    theta_pi_0fold_MTPLUS.append(sfs0.theta_pi())\n",
    "    theta_pi_4fold_MTPLUS.append(sfs4.theta_pi())\n",
    "    theta_pi_total_MTPLUS.append(sfst.theta_pi())\n",
    "    theta_w_0fold_MTPLUS.append(sfs0.theta_w())\n",
    "    theta_w_4fold_MTPLUS.append(sfs4.theta_w())\n",
    "    theta_w_total_MTPLUS.append(sfst.theta_w())\n",
    "    sfs_4fold_MTPLUS.append(sfs4.sfs)\n",
    "    sfs_0fold_MTPLUS.append(sfs0.sfs)\n",
    "    sfs_t_MTPLUS.append(sfst.sfs)   \n",
    "    ###################################\n",
    "    ## MTMINUS\n",
    "    sfst,sfs0,sfs4 = alignment2sfs_w_degen(aln, 8, degen_map, samples=mtMinus_sampleIDs, start=None, end=None)\n",
    "    ##### record ALL results\n",
    "    alleles_MTMINUS.append(sfst.alleles)\n",
    "    sites_MTMINUS.append(sfst.sites())\n",
    "    invariant_MTMINUS.append(sfst.invariant())\n",
    "    variant_MTMINUS.append(sfst.variant())\n",
    "    theta_pi_0fold_MTMINUS.append(sfs0.theta_pi())\n",
    "    theta_pi_4fold_MTMINUS.append(sfs4.theta_pi())\n",
    "    theta_pi_total_MTMINUS.append(sfst.theta_pi())\n",
    "    theta_w_0fold_MTMINUS.append(sfs0.theta_w())\n",
    "    theta_w_4fold_MTMINUS.append(sfs4.theta_w())\n",
    "    theta_w_total_MTMINUS.append(sfst.theta_w())\n",
    "    sfs_4fold_MTMINUS.append(sfs4.sfs)\n",
    "    sfs_0fold_MTMINUS.append(sfs0.sfs)\n",
    "    sfs_t_MTMINUS.append(sfst.sfs)   \n",
    "    ###################################\n",
    "    ## population stats\n",
    "    pop_stats= population_poly(aln, 12, degen_map, mtPlus_sampleIDs, mtMinus_sampleIDs, start=None, end=None)\n",
    "    private_MTPLUS.append(  pop_stats['private1'])\n",
    "    private_MTMINUS.append( pop_stats['private2'])\n",
    "    shared.append(pop_stats['shared'])\n",
    "    fixed.append(pop_stats['fixed'])\n",
    "    Fst.append(pop_stats['fst'])\n",
    "    #######\n",
    "    ## GC_4fold\n",
    "    try:\n",
    "        gc =len([1 for i,j in zip(special_consensus(aln),degen_map) if i in 'GCgc' and j =='4'])/\n",
    "            len([1 for i,j in zip(special_consensus(aln),degen_map) if i in 'AaTtGCgc' and j =='4'])\n",
    "    except ZeroDivisionError:\n",
    "        gc = 'N/A'\n",
    "    GC4.append(gc)\n",
    "\n",
    "#########################################################\n",
    "#### DF the results #####################################\n",
    "#########################################################\n",
    "nonMTlocus_results = pd.DataFrame(\n",
    "    {'gene' : genes,\n",
    "     'mtPlus_coords' : mtPlus_coords,\n",
    "     'mtMinus_coords' : mtMinus_coords,\n",
    "     'alleles_ALL' : alleles_ALL, \n",
    "     'sites_ALL' : sites_ALL, \n",
    "     'invariant_ALL' : invariant_ALL, \n",
    "     'variant_ALL' : variant_ALL, \n",
    "     'theta_pi_0fold_ALL' : theta_pi_0fold_ALL, \n",
    "     'theta_pi_4fold_ALL' : theta_pi_4fold_ALL, \n",
    "     'theta_pi_total_ALL' : theta_pi_total_ALL, \n",
    "     'theta_w_0fold_ALL' : theta_w_0fold_ALL, \n",
    "     'theta_w_4fold_ALL' : theta_w_4fold_ALL, \n",
    "     'theta_w_total_ALL' : theta_w_total_ALL, \n",
    "     'sfs_4fold_ALL' : sfs_4fold_ALL, \n",
    "     'sfs_0fold_ALL' : sfs_0fold_ALL, \n",
    "     'sfs_t_ALL' : sfs_t_ALL,    \n",
    "     'alleles_MTPLUS' : alleles_MTPLUS, \n",
    "     'sites_MTPLUS' : sites_MTPLUS, \n",
    "     'invariant_MTPLUS' : invariant_MTPLUS, \n",
    "     'variant_MTPLUS' : variant_MTPLUS, \n",
    "     'theta_pi_0fold_MTPLUS' : theta_pi_0fold_MTPLUS, \n",
    "     'theta_pi_4fold_MTPLUS' : theta_pi_4fold_MTPLUS, \n",
    "     'theta_pi_total_MTPLUS' : theta_pi_total_MTPLUS, \n",
    "     'theta_w_0fold_MTPLUS' : theta_w_0fold_MTPLUS, \n",
    "     'theta_w_4fold_MTPLUS' : theta_w_4fold_MTPLUS, \n",
    "     'theta_w_total_MTPLUS' : theta_w_total_MTPLUS, \n",
    "     'sfs_4fold_MTPLUS' : sfs_4fold_MTPLUS, \n",
    "     'sfs_0fold_MTPLUS' : sfs_0fold_MTPLUS, \n",
    "     'sfs_t_MTPLUS' : sfs_t_MTPLUS,         \n",
    "     'alleles_MTMINUS' : alleles_MTMINUS, \n",
    "     'sites_MTMINUS' : sites_MTMINUS, \n",
    "     'invariant_MTMINUS' : invariant_MTMINUS, \n",
    "     'variant_MTMINUS' : variant_MTMINUS, \n",
    "     'theta_pi_0fold_MTMINUS' : theta_pi_0fold_MTMINUS, \n",
    "     'theta_pi_4fold_MTMINUS' : theta_pi_4fold_MTMINUS, \n",
    "     'theta_pi_total_MTMINUS' : theta_pi_total_MTMINUS, \n",
    "     'theta_w_0fold_MTMINUS' : theta_w_0fold_MTMINUS, \n",
    "     'theta_w_4fold_MTMINUS' : theta_w_4fold_MTMINUS, \n",
    "     'theta_w_total_MTMINUS' : theta_w_total_MTMINUS, \n",
    "     'sfs_4fold_MTMINUS' : sfs_4fold_MTMINUS, \n",
    "     'sfs_0fold_MTMINUS' : sfs_0fold_MTMINUS, \n",
    "     'sfs_t_MTMINUS' : sfs_t_MTMINUS, \n",
    "     'private_MTPLUS' : private_MTPLUS, \n",
    "     'private_MTMINUS' : private_MTMINUS, \n",
    "     'shared' : shared, \n",
    "     'fixed' : fixed, \n",
    "     'Fst' : Fst, \n",
    "     'GC4' : GC4\n",
    "    }) \n",
    "\n",
    "\n",
    "nonMTlocus_results.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "nonMTlocus_results.to_csv(\"polymorphism_all.NonMTLocus.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Gametologous Regions        \n",
    "---\n",
    "\n",
    "\n",
    "Here I go through the gametolog alignment that has been pinned together to map to the mtPlus allele \n",
    "and I calculate stats for silent sites including\n",
    " - 4 fold degneration\n",
    " - intron\n",
    " - intergenic\n",
    "\n",
    "I also calculate a NOT silent bin which includes UTR/0-fold/2-fold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Fst</th>\n",
       "      <th>GCsilent</th>\n",
       "      <th>alleles_ALL</th>\n",
       "      <th>alleles_MTMINUS</th>\n",
       "      <th>alleles_MTPLUS</th>\n",
       "      <th>fixed</th>\n",
       "      <th>gene</th>\n",
       "      <th>invariant_ALL</th>\n",
       "      <th>invariant_MTMINUS</th>\n",
       "      <th>invariant_MTPLUS</th>\n",
       "      <th>...</th>\n",
       "      <th>theta_w_notSilent_MTPLUS</th>\n",
       "      <th>theta_w_silent_ALL</th>\n",
       "      <th>theta_w_silent_MTMINUS</th>\n",
       "      <th>theta_w_silent_MTPLUS</th>\n",
       "      <th>theta_w_total_ALL</th>\n",
       "      <th>theta_w_total_MTMINUS</th>\n",
       "      <th>theta_w_total_MTPLUS</th>\n",
       "      <th>variant_ALL</th>\n",
       "      <th>variant_MTMINUS</th>\n",
       "      <th>variant_MTPLUS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.429806</td>\n",
       "      <td>0.657755</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>5525</td>\n",
       "      <td>mt_aligned_all</td>\n",
       "      <td>395496</td>\n",
       "      <td>400640</td>\n",
       "      <td>415016</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003757</td>\n",
       "      <td>0.016504</td>\n",
       "      <td>0.012225</td>\n",
       "      <td>0.007162</td>\n",
       "      <td>0.013267</td>\n",
       "      <td>0.009707</td>\n",
       "      <td>0.005548</td>\n",
       "      <td>19231</td>\n",
       "      <td>11313</td>\n",
       "      <td>6354</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows  48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Fst  GCsilent  alleles_ALL  alleles_MTMINUS  alleles_MTPLUS  fixed  \\\n",
       "0  0.429806  0.657755           19               10               9   5525   \n",
       "\n",
       "             gene  invariant_ALL  invariant_MTMINUS  invariant_MTPLUS  \\\n",
       "0  mt_aligned_all         395496             400640            415016   \n",
       "\n",
       "       ...       theta_w_notSilent_MTPLUS theta_w_silent_ALL  \\\n",
       "0      ...                       0.003757           0.016504   \n",
       "\n",
       "   theta_w_silent_MTMINUS  theta_w_silent_MTPLUS theta_w_total_ALL  \\\n",
       "0                0.012225               0.007162          0.013267   \n",
       "\n",
       "  theta_w_total_MTMINUS theta_w_total_MTPLUS variant_ALL variant_MTMINUS  \\\n",
       "0              0.009707             0.005548       19231           11313   \n",
       "\n",
       "  variant_MTPLUS  \n",
       "0           6354  \n",
       "\n",
       "[1 rows x 48 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Target Genotype Lists\n",
    "mtPlus_Genotypes  = \"CC2936 CC2937 CC3060 CC3064 CC3065 CC3068 CC3071 CC3076 CC3086\".split()\n",
    "mtMinus_Genotypes = \"CC2935 CC2938 CC3059 CC3061 CC3062 CC3063 CC3073 CC3075 CC3079 CC3084\".split()\n",
    "\n",
    "\n",
    "#### STATS TO COLLECT \n",
    "genes = []\n",
    "mtPlus_coords = []\n",
    "mtMinus_coords = []\n",
    "alleles_ALL = []\n",
    "sites_ALL = []\n",
    "invariant_ALL = []\n",
    "variant_ALL = []\n",
    "theta_pi_notSilent_ALL = []\n",
    "theta_pi_silent_ALL = []\n",
    "theta_pi_total_ALL = []\n",
    "theta_w_notSilent_ALL = []\n",
    "theta_w_silent_ALL = []\n",
    "theta_w_total_ALL = []\n",
    "sfs_silent_ALL = []\n",
    "sfs_notSilent_ALL = []\n",
    "sfs_t_ALL = []\n",
    "alleles_MTPLUS = []\n",
    "sites_MTPLUS = []\n",
    "invariant_MTPLUS = []\n",
    "variant_MTPLUS = []\n",
    "theta_pi_notSilent_MTPLUS = []\n",
    "theta_pi_silent_MTPLUS = []\n",
    "theta_pi_total_MTPLUS = []\n",
    "theta_w_notSilent_MTPLUS = []\n",
    "theta_w_silent_MTPLUS = []\n",
    "theta_w_total_MTPLUS = []\n",
    "sfs_silent_MTPLUS = []\n",
    "sfs_notSilent_MTPLUS = []\n",
    "sfs_t_MTPLUS = []\n",
    "alleles_MTMINUS = []\n",
    "sites_MTMINUS = []\n",
    "invariant_MTMINUS = []\n",
    "variant_MTMINUS = []\n",
    "theta_pi_notSilent_MTMINUS = []\n",
    "theta_pi_silent_MTMINUS = []\n",
    "theta_pi_total_MTMINUS = []\n",
    "theta_w_notSilent_MTMINUS = []\n",
    "theta_w_silent_MTMINUS = []\n",
    "theta_w_total_MTMINUS = []\n",
    "sfs_silent_MTMINUS = []\n",
    "sfs_notSilent_MTMINUS = []\n",
    "sfs_t_MTMINUS = []\n",
    "private_MTPLUS = []\n",
    "private_MTMINUS = []\n",
    "shared = []\n",
    "fixed = []\n",
    "Fst = []\n",
    "GCsilent = []\n",
    "\n",
    "\n",
    "# To calculate silent/vs non-silent stats I can manipulate the degen map as follows:\n",
    "# - 4 = all silent sites -intron, intergenic, 4-fold\n",
    "# I need to use the annotation table for the region   \n",
    "# 298298-943474 (0-based python style)\n",
    "import annotation_table, pysam\n",
    "annotation_table_file = \"/scratch/research/references/chlamydomonas/5.3_chlamy_w_organelles_mt_minus/annotation/concatenated_GFF/annotation_table.txt.gz\"\n",
    "at = pysam.Tabixfile(annotation_table_file)\n",
    "special_degen_map = ''\n",
    "for l in  at.fetch('chromosome_6',298298,943474): #note used 0-based non inclusive coords\n",
    "    a = annotation_table.annotation_line(l)\n",
    "    if a.fold4 =='1' or a.intergenic =='1' or a.intronic =='1': # all silent sites are marked as 4-fold\n",
    "        special_degen_map+='4'\n",
    "    else:\n",
    "        special_degen_map+='0'\n",
    "\n",
    "wtfs = glob.glob(\"../../../data/aligned-fastas/mt_aligned_all.fasta\")\n",
    "for a in wtfs:\n",
    "    gene = a.split(\"/\")[-1].split(\".\")[0]\n",
    "    genes.append(gene)\n",
    "    aln = AlignIO.read(open(a), 'fasta')\n",
    "    mtPlus_sampleIDs  = [s.id for s in aln if s.id.split(\"|\")[0] in mtPlus_Genotypes]\n",
    "    mtMinus_sampleIDs = [s.id for s in aln if s.id.split(\"|\")[0] in mtMinus_Genotypes]\n",
    "    mtPlus_coords.append('gametologs_pluscoords') \n",
    "    mtMinus_coords.append('gametologs_pluscoords')\n",
    "    degen_map = special_degen_map\n",
    "    sfst,sfs0,sfs4 = alignment2sfs_w_degen(aln, 12, degen_map, samples=None, start=None, end=None)\n",
    "    ##### record ALL results\n",
    "    alleles_ALL.append(sfst.alleles)\n",
    "    sites_ALL.append(sfst.sites())\n",
    "    invariant_ALL.append(sfst.invariant())\n",
    "    variant_ALL.append(sfst.variant())\n",
    "    theta_pi_notSilent_ALL.append(sfs0.theta_pi())\n",
    "    theta_pi_silent_ALL.append(sfs4.theta_pi())\n",
    "    theta_pi_total_ALL.append(sfst.theta_pi())\n",
    "    theta_w_notSilent_ALL.append(sfs0.theta_w())\n",
    "    theta_w_silent_ALL.append(sfs4.theta_w())\n",
    "    theta_w_total_ALL.append(sfst.theta_w())\n",
    "    sfs_silent_ALL.append(sfs4.sfs)\n",
    "    sfs_notSilent_ALL.append(sfs0.sfs)\n",
    "    sfs_t_ALL.append(sfst.sfs)\n",
    "    ###################################\n",
    "    ## MTPLUS\n",
    "    sfst,sfs0,sfs4 = alignment2sfs_w_degen(aln, 8, degen_map, samples=mtPlus_sampleIDs, start=None, end=None)\n",
    "    ##### record ALL results\n",
    "    alleles_MTPLUS.append(sfst.alleles)\n",
    "    sites_MTPLUS.append(sfst.sites())\n",
    "    invariant_MTPLUS.append(sfst.invariant())\n",
    "    variant_MTPLUS.append(sfst.variant())\n",
    "    theta_pi_notSilent_MTPLUS.append(sfs0.theta_pi())\n",
    "    theta_pi_silent_MTPLUS.append(sfs4.theta_pi())\n",
    "    theta_pi_total_MTPLUS.append(sfst.theta_pi())\n",
    "    theta_w_notSilent_MTPLUS.append(sfs0.theta_w())\n",
    "    theta_w_silent_MTPLUS.append(sfs4.theta_w())\n",
    "    theta_w_total_MTPLUS.append(sfst.theta_w())\n",
    "    sfs_silent_MTPLUS.append(sfs4.sfs)\n",
    "    sfs_notSilent_MTPLUS.append(sfs0.sfs)\n",
    "    sfs_t_MTPLUS.append(sfst.sfs)   \n",
    "    ###################################\n",
    "    ## MTMINUS\n",
    "    sfst,sfs0,sfs4 = alignment2sfs_w_degen(aln, 8, degen_map, samples=mtMinus_sampleIDs, start=None, end=None)\n",
    "    ##### record ALL results\n",
    "    alleles_MTMINUS.append(sfst.alleles)\n",
    "    sites_MTMINUS.append(sfst.sites())\n",
    "    invariant_MTMINUS.append(sfst.invariant())\n",
    "    variant_MTMINUS.append(sfst.variant())\n",
    "    theta_pi_notSilent_MTMINUS.append(sfs0.theta_pi())\n",
    "    theta_pi_silent_MTMINUS.append(sfs4.theta_pi())\n",
    "    theta_pi_total_MTMINUS.append(sfst.theta_pi())\n",
    "    theta_w_notSilent_MTMINUS.append(sfs0.theta_w())\n",
    "    theta_w_silent_MTMINUS.append(sfs4.theta_w())\n",
    "    theta_w_total_MTMINUS.append(sfst.theta_w())\n",
    "    sfs_silent_MTMINUS.append(sfs4.sfs)\n",
    "    sfs_notSilent_MTMINUS.append(sfs0.sfs)\n",
    "    sfs_t_MTMINUS.append(sfst.sfs)   \n",
    "    ###################################\n",
    "    ## population stats\n",
    "    pop_stats= population_poly(aln, 12, degen_map, mtPlus_sampleIDs, mtMinus_sampleIDs, start=None, end=None)\n",
    "    private_MTPLUS.append(  pop_stats['private1'])\n",
    "    private_MTMINUS.append( pop_stats['private2'])\n",
    "    shared.append(pop_stats['shared'])\n",
    "    fixed.append(pop_stats['fixed'])\n",
    "    Fst.append(pop_stats['fst'])\n",
    "    #######\n",
    "    ## GC_silent\n",
    "    GCsilent.append(len([1 for i,j in zip(special_consensus(aln),degen_map) if i in 'GCgc' and j =='4'])/\n",
    "               len([1 for i,j in zip(special_consensus(aln),degen_map) if i in 'AaTtGCgc' and j =='4']))\n",
    "\n",
    "    \n",
    "#########################################################\n",
    "#### DF the results #####################################\n",
    "#########################################################\n",
    "gametolog_results = pd.DataFrame(\n",
    "    {'gene' : genes,\n",
    "     'mtPlus_coords' : mtPlus_coords,\n",
    "     'mtMinus_coords' : mtMinus_coords,\n",
    "     'alleles_ALL' : alleles_ALL, \n",
    "     'sites_ALL' : sites_ALL, \n",
    "     'invariant_ALL' : invariant_ALL, \n",
    "     'variant_ALL' : variant_ALL, \n",
    "     'theta_pi_notSilent_ALL' : theta_pi_notSilent_ALL, \n",
    "     'theta_pi_silent_ALL' : theta_pi_silent_ALL, \n",
    "     'theta_pi_total_ALL' : theta_pi_total_ALL, \n",
    "     'theta_w_notSilent_ALL' : theta_w_notSilent_ALL, \n",
    "     'theta_w_silent_ALL' : theta_w_silent_ALL, \n",
    "     'theta_w_total_ALL' : theta_w_total_ALL, \n",
    "     'sfs_silent_ALL' : sfs_silent_ALL, \n",
    "     'sfs_notSilent_ALL' : sfs_notSilent_ALL, \n",
    "     'sfs_t_ALL' : sfs_t_ALL,    \n",
    "     'alleles_MTPLUS' : alleles_MTPLUS, \n",
    "     'sites_MTPLUS' : sites_MTPLUS, \n",
    "     'invariant_MTPLUS' : invariant_MTPLUS, \n",
    "     'variant_MTPLUS' : variant_MTPLUS, \n",
    "     'theta_pi_notSilent_MTPLUS' : theta_pi_notSilent_MTPLUS, \n",
    "     'theta_pi_silent_MTPLUS' : theta_pi_silent_MTPLUS, \n",
    "     'theta_pi_total_MTPLUS' : theta_pi_total_MTPLUS, \n",
    "     'theta_w_notSilent_MTPLUS' : theta_w_notSilent_MTPLUS, \n",
    "     'theta_w_silent_MTPLUS' : theta_w_silent_MTPLUS, \n",
    "     'theta_w_total_MTPLUS' : theta_w_total_MTPLUS, \n",
    "     'sfs_silent_MTPLUS' : sfs_silent_MTPLUS, \n",
    "     'sfs_notSilent_MTPLUS' : sfs_notSilent_MTPLUS, \n",
    "     'sfs_t_MTPLUS' : sfs_t_MTPLUS,         \n",
    "     'alleles_MTMINUS' : alleles_MTMINUS, \n",
    "     'sites_MTMINUS' : sites_MTMINUS, \n",
    "     'invariant_MTMINUS' : invariant_MTMINUS, \n",
    "     'variant_MTMINUS' : variant_MTMINUS, \n",
    "     'theta_pi_notSilent_MTMINUS' : theta_pi_notSilent_MTMINUS, \n",
    "     'theta_pi_silent_MTMINUS' : theta_pi_silent_MTMINUS, \n",
    "     'theta_pi_total_MTMINUS' : theta_pi_total_MTMINUS, \n",
    "     'theta_w_notSilent_MTMINUS' : theta_w_notSilent_MTMINUS, \n",
    "     'theta_w_silent_MTMINUS' : theta_w_silent_MTMINUS, \n",
    "     'theta_w_total_MTMINUS' : theta_w_total_MTMINUS, \n",
    "     'sfs_silent_MTMINUS' : sfs_silent_MTMINUS, \n",
    "     'sfs_notSilent_MTMINUS' : sfs_notSilent_MTMINUS, \n",
    "     'sfs_t_MTMINUS' : sfs_t_MTMINUS, \n",
    "     'private_MTPLUS' : private_MTPLUS, \n",
    "     'private_MTMINUS' : private_MTMINUS, \n",
    "     'shared' : shared, \n",
    "     'fixed' : fixed, \n",
    "     'Fst' : Fst, \n",
    "     'GCsilent' : GCsilent\n",
    "    })\n",
    "\n",
    "\n",
    "gametolog_results.to_csv(\"polymorphism_all.gametolog.csv\", index=False)\n",
    "gametolog_results.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Gametolog w Intron and Intergenic vs other\n",
    "---\n",
    "\n",
    "\n",
    "Here I go through the gametolog alignment that has been pinned together to map to the mtPlus allele \n",
    "and I calculate stats for intron and intergenic sites and total sites\n",
    "\n",
    "** Although 0-fold is reported there are no data in this category because it was masked **\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Fst</th>\n",
       "      <th>GC_IandI</th>\n",
       "      <th>alleles_ALL</th>\n",
       "      <th>alleles_MTMINUS</th>\n",
       "      <th>alleles_MTPLUS</th>\n",
       "      <th>fixed</th>\n",
       "      <th>gene</th>\n",
       "      <th>invariant_ALL</th>\n",
       "      <th>invariant_MTMINUS</th>\n",
       "      <th>invariant_MTPLUS</th>\n",
       "      <th>...</th>\n",
       "      <th>theta_w_0fold_MTPLUS</th>\n",
       "      <th>theta_w_IandI_ALL</th>\n",
       "      <th>theta_w_IandI_MTMINUS</th>\n",
       "      <th>theta_w_IandI_MTPLUS</th>\n",
       "      <th>theta_w_total_ALL</th>\n",
       "      <th>theta_w_total_MTMINUS</th>\n",
       "      <th>theta_w_total_MTPLUS</th>\n",
       "      <th>variant_ALL</th>\n",
       "      <th>variant_MTMINUS</th>\n",
       "      <th>variant_MTPLUS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.429806</td>\n",
       "      <td>0.633101</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>5525</td>\n",
       "      <td>mt_aligned_all</td>\n",
       "      <td>395496</td>\n",
       "      <td>400640</td>\n",
       "      <td>415016</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.017022</td>\n",
       "      <td>0.012209</td>\n",
       "      <td>0.007242</td>\n",
       "      <td>0.013267</td>\n",
       "      <td>0.009707</td>\n",
       "      <td>0.005548</td>\n",
       "      <td>19231</td>\n",
       "      <td>11313</td>\n",
       "      <td>6354</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows  48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Fst  GC_IandI  alleles_ALL  alleles_MTMINUS  alleles_MTPLUS  fixed  \\\n",
       "0  0.429806  0.633101           19               10               9   5525   \n",
       "\n",
       "             gene  invariant_ALL  invariant_MTMINUS  invariant_MTPLUS  \\\n",
       "0  mt_aligned_all         395496             400640            415016   \n",
       "\n",
       "       ...       theta_w_0fold_MTPLUS theta_w_IandI_ALL  \\\n",
       "0      ...                        NaN          0.017022   \n",
       "\n",
       "   theta_w_IandI_MTMINUS  theta_w_IandI_MTPLUS theta_w_total_ALL  \\\n",
       "0               0.012209              0.007242          0.013267   \n",
       "\n",
       "  theta_w_total_MTMINUS theta_w_total_MTPLUS variant_ALL variant_MTMINUS  \\\n",
       "0              0.009707             0.005548       19231           11313   \n",
       "\n",
       "  variant_MTPLUS  \n",
       "0           6354  \n",
       "\n",
       "[1 rows x 48 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#### STATS TO COLLECT \n",
    "genes = []\n",
    "mtPlus_coords = []\n",
    "mtMinus_coords = []\n",
    "alleles_ALL = []\n",
    "sites_ALL = []\n",
    "invariant_ALL = []\n",
    "variant_ALL = []\n",
    "theta_pi_0fold_ALL = []\n",
    "theta_pi_IandI_ALL = []\n",
    "theta_pi_total_ALL = []\n",
    "theta_w_0fold_ALL = []\n",
    "theta_w_IandI_ALL = []\n",
    "theta_w_total_ALL = []\n",
    "sfs_IandI_ALL = []\n",
    "sfs_0fold_ALL = []\n",
    "sfs_t_ALL = []\n",
    "alleles_MTPLUS = []\n",
    "sites_MTPLUS = []\n",
    "invariant_MTPLUS = []\n",
    "variant_MTPLUS = []\n",
    "theta_pi_0fold_MTPLUS = []\n",
    "theta_pi_IandI_MTPLUS = []\n",
    "theta_pi_total_MTPLUS = []\n",
    "theta_w_0fold_MTPLUS = []\n",
    "theta_w_IandI_MTPLUS = []\n",
    "theta_w_total_MTPLUS = []\n",
    "sfs_IandI_MTPLUS = []\n",
    "sfs_0fold_MTPLUS = []\n",
    "sfs_t_MTPLUS = []\n",
    "alleles_MTMINUS = []\n",
    "sites_MTMINUS = []\n",
    "invariant_MTMINUS = []\n",
    "variant_MTMINUS = []\n",
    "theta_pi_0fold_MTMINUS = []\n",
    "theta_pi_IandI_MTMINUS = []\n",
    "theta_pi_total_MTMINUS = []\n",
    "theta_w_0fold_MTMINUS = []\n",
    "theta_w_IandI_MTMINUS = []\n",
    "theta_w_total_MTMINUS = []\n",
    "sfs_IandI_MTMINUS = []\n",
    "sfs_0fold_MTMINUS = []\n",
    "sfs_t_MTMINUS = []\n",
    "private_MTPLUS = []\n",
    "private_MTMINUS = []\n",
    "shared = []\n",
    "fixed = []\n",
    "Fst = []\n",
    "GC_IandI = []\n",
    "\n",
    "\n",
    "# To calculate silent/vs non-silent stats I can manipulate the degen map as follows:\n",
    "# - 4 = all silent sites -intron, intergenic, 4-fold\n",
    "# I need to use the annotation table for the region   \n",
    "# 298298-943474 (0-based python style)\n",
    "import annotation_table, pysam\n",
    "annotation_table_file = \"/scratch/research/references/chlamydomonas/5.3_chlamy_w_organelles_mt_minus/annotation/concatenated_GFF/annotation_table.txt.gz\"\n",
    "at = pysam.Tabixfile(annotation_table_file)\n",
    "special_degen_map = ''\n",
    "for l in  at.fetch('chromosome_6',298298,943474): #note used 0-based non inclusive coords\n",
    "    a = annotation_table.annotation_line(l)\n",
    "    if a.intergenic =='1' or a.intronic =='1':\n",
    "        special_degen_map+='4'\n",
    "    else:\n",
    "        special_degen_map+='2' # this effectively eliminates 0-fold stats\n",
    "\n",
    "wtfs = glob.glob(\"../../../data/aligned-fastas/mt_aligned_all.fasta\")\n",
    "for a in wtfs:\n",
    "    gene = a.split(\"/\")[-1].split(\".\")[0]\n",
    "    genes.append(gene)\n",
    "    aln = AlignIO.read(open(a), 'fasta')\n",
    "    mtPlus_sampleIDs  = [s.id for s in aln if s.id.split(\"|\")[0] in mtPlus_Genotypes]\n",
    "    mtMinus_sampleIDs = [s.id for s in aln if s.id.split(\"|\")[0] in mtMinus_Genotypes]\n",
    "    mtPlus_coords.append('gametologs_pluscoords_intron_intergenic') \n",
    "    mtMinus_coords.append('gametologs_pluscoords_intron_intergenic')\n",
    "    degen_map = special_degen_map\n",
    "    sfst,sfs0,sfs4 = alignment2sfs_w_degen(aln, 12, degen_map, samples=None, start=None, end=None)\n",
    "    ##### record ALL results\n",
    "    alleles_ALL.append(sfst.alleles)\n",
    "    sites_ALL.append(sfst.sites())\n",
    "    invariant_ALL.append(sfst.invariant())\n",
    "    variant_ALL.append(sfst.variant())\n",
    "    theta_pi_0fold_ALL.append(sfs0.theta_pi())\n",
    "    theta_pi_IandI_ALL.append(sfs4.theta_pi())\n",
    "    theta_pi_total_ALL.append(sfst.theta_pi())\n",
    "    theta_w_0fold_ALL.append(sfs0.theta_w())\n",
    "    theta_w_IandI_ALL.append(sfs4.theta_w())\n",
    "    theta_w_total_ALL.append(sfst.theta_w())\n",
    "    sfs_IandI_ALL.append(sfs4.sfs)\n",
    "    sfs_0fold_ALL.append(sfs0.sfs)\n",
    "    sfs_t_ALL.append(sfst.sfs)\n",
    "    ###################################\n",
    "    ## MTPLUS\n",
    "    sfst,sfs0,sfs4 = alignment2sfs_w_degen(aln, 8, degen_map, samples=mtPlus_sampleIDs, start=None, end=None)\n",
    "    ##### record ALL results\n",
    "    alleles_MTPLUS.append(sfst.alleles)\n",
    "    sites_MTPLUS.append(sfst.sites())\n",
    "    invariant_MTPLUS.append(sfst.invariant())\n",
    "    variant_MTPLUS.append(sfst.variant())\n",
    "    theta_pi_0fold_MTPLUS.append(sfs0.theta_pi())\n",
    "    theta_pi_IandI_MTPLUS.append(sfs4.theta_pi())\n",
    "    theta_pi_total_MTPLUS.append(sfst.theta_pi())\n",
    "    theta_w_0fold_MTPLUS.append(sfs0.theta_w())\n",
    "    theta_w_IandI_MTPLUS.append(sfs4.theta_w())\n",
    "    theta_w_total_MTPLUS.append(sfst.theta_w())\n",
    "    sfs_IandI_MTPLUS.append(sfs4.sfs)\n",
    "    sfs_0fold_MTPLUS.append(sfs0.sfs)\n",
    "    sfs_t_MTPLUS.append(sfst.sfs)   \n",
    "    ###################################\n",
    "    ## MTMINUS\n",
    "    sfst,sfs0,sfs4 = alignment2sfs_w_degen(aln, 8, degen_map, samples=mtMinus_sampleIDs, start=None, end=None)\n",
    "    ##### record ALL results\n",
    "    alleles_MTMINUS.append(sfst.alleles)\n",
    "    sites_MTMINUS.append(sfst.sites())\n",
    "    invariant_MTMINUS.append(sfst.invariant())\n",
    "    variant_MTMINUS.append(sfst.variant())\n",
    "    theta_pi_0fold_MTMINUS.append(sfs0.theta_pi())\n",
    "    theta_pi_IandI_MTMINUS.append(sfs4.theta_pi())\n",
    "    theta_pi_total_MTMINUS.append(sfst.theta_pi())\n",
    "    theta_w_0fold_MTMINUS.append(sfs0.theta_w())\n",
    "    theta_w_IandI_MTMINUS.append(sfs4.theta_w())\n",
    "    theta_w_total_MTMINUS.append(sfst.theta_w())\n",
    "    sfs_IandI_MTMINUS.append(sfs4.sfs)\n",
    "    sfs_0fold_MTMINUS.append(sfs0.sfs)\n",
    "    sfs_t_MTMINUS.append(sfst.sfs)   \n",
    "    ###################################\n",
    "    ## population stats\n",
    "    pop_stats= population_poly(aln, 12, degen_map, mtPlus_sampleIDs, mtMinus_sampleIDs, start=None, end=None)\n",
    "    private_MTPLUS.append(  pop_stats['private1'])\n",
    "    private_MTMINUS.append( pop_stats['private2'])\n",
    "    shared.append(pop_stats['shared'])\n",
    "    fixed.append(pop_stats['fixed'])\n",
    "    Fst.append(pop_stats['fst'])\n",
    "    #######\n",
    "    ## GC_IandI\n",
    "    GC_IandI.append(len([1 for i,j in zip(special_consensus(aln),degen_map) if i in 'GCgc' and j =='4'])/\n",
    "               len([1 for i,j in zip(special_consensus(aln),degen_map) if i in 'AaTtGCgc' and j =='4']))\n",
    "\n",
    "    \n",
    "#########################################################\n",
    "#### DF the results #####################################\n",
    "#########################################################\n",
    "gametolog_intron_intergenic_results = pd.DataFrame(\n",
    "    {'gene' : genes,\n",
    "     'mtPlus_coords' : mtPlus_coords,\n",
    "     'mtMinus_coords' : mtMinus_coords,\n",
    "     'alleles_ALL' : alleles_ALL, \n",
    "     'sites_ALL' : sites_ALL, \n",
    "     'invariant_ALL' : invariant_ALL, \n",
    "     'variant_ALL' : variant_ALL, \n",
    "     'theta_pi_0fold_ALL' : theta_pi_0fold_ALL, \n",
    "     'theta_pi_IandI_ALL' : theta_pi_IandI_ALL, \n",
    "     'theta_pi_total_ALL' : theta_pi_total_ALL, \n",
    "     'theta_w_0fold_ALL' : theta_w_0fold_ALL, \n",
    "     'theta_w_IandI_ALL' : theta_w_IandI_ALL, \n",
    "     'theta_w_total_ALL' : theta_w_total_ALL, \n",
    "     'sfs_IandI_ALL' : sfs_IandI_ALL, \n",
    "     'sfs_0fold_ALL' : sfs_0fold_ALL, \n",
    "     'sfs_t_ALL' : sfs_t_ALL,    \n",
    "     'alleles_MTPLUS' : alleles_MTPLUS, \n",
    "     'sites_MTPLUS' : sites_MTPLUS, \n",
    "     'invariant_MTPLUS' : invariant_MTPLUS, \n",
    "     'variant_MTPLUS' : variant_MTPLUS, \n",
    "     'theta_pi_0fold_MTPLUS' : theta_pi_0fold_MTPLUS, \n",
    "     'theta_pi_IandI_MTPLUS' : theta_pi_IandI_MTPLUS, \n",
    "     'theta_pi_total_MTPLUS' : theta_pi_total_MTPLUS, \n",
    "     'theta_w_0fold_MTPLUS' : theta_w_0fold_MTPLUS, \n",
    "     'theta_w_IandI_MTPLUS' : theta_w_IandI_MTPLUS, \n",
    "     'theta_w_total_MTPLUS' : theta_w_total_MTPLUS, \n",
    "     'sfs_IandI_MTPLUS' : sfs_IandI_MTPLUS, \n",
    "     'sfs_0fold_MTPLUS' : sfs_0fold_MTPLUS, \n",
    "     'sfs_t_MTPLUS' : sfs_t_MTPLUS,         \n",
    "     'alleles_MTMINUS' : alleles_MTMINUS, \n",
    "     'sites_MTMINUS' : sites_MTMINUS, \n",
    "     'invariant_MTMINUS' : invariant_MTMINUS, \n",
    "     'variant_MTMINUS' : variant_MTMINUS, \n",
    "     'theta_pi_0fold_MTMINUS' : theta_pi_0fold_MTMINUS, \n",
    "     'theta_pi_IandI_MTMINUS' : theta_pi_IandI_MTMINUS, \n",
    "     'theta_pi_total_MTMINUS' : theta_pi_total_MTMINUS, \n",
    "     'theta_w_0fold_MTMINUS' : theta_w_0fold_MTMINUS, \n",
    "     'theta_w_IandI_MTMINUS' : theta_w_IandI_MTMINUS, \n",
    "     'theta_w_total_MTMINUS' : theta_w_total_MTMINUS, \n",
    "     'sfs_IandI_MTMINUS' : sfs_IandI_MTMINUS, \n",
    "     'sfs_0fold_MTMINUS' : sfs_0fold_MTMINUS, \n",
    "     'sfs_t_MTMINUS' : sfs_t_MTMINUS, \n",
    "     'private_MTPLUS' : private_MTPLUS, \n",
    "     'private_MTMINUS' : private_MTMINUS, \n",
    "     'shared' : shared, \n",
    "     'fixed' : fixed, \n",
    "     'Fst' : Fst, \n",
    "     'GC_IandI' : GC_IandI\n",
    "    })\n",
    "\n",
    "\n",
    "gametolog_intron_intergenic_results.to_csv(\"polymorphism_all.gametolog_intron_intergenic.csv\", index=False)\n",
    "gametolog_intron_intergenic_results.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7299 2237 5525 4170\n"
     ]
    }
   ],
   "source": [
    "print(gametolog_results.private_MTMINUS[0],\n",
    "gametolog_results.private_MTPLUS[0],\n",
    "gametolog_results.fixed[0],\n",
    "gametolog_results.shared[0],)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Non-gametologous regions\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['../../../data/aligned-fastas/plus_non_gametolog.fasta', '../../../data/aligned-fastas/minus_non_gametolog.fasta']\n",
      ">CC3076\n",
      ">CC3061\n"
     ]
    }
   ],
   "source": [
    "# Target Genotype Lists\n",
    "mtPlus_Genotypes = \"CC2936 CC2937 CC3060 CC3064 CC3065 CC3068 CC3071 CC3076 CC3086\".split()\n",
    "mtMinus_Genotypes = \"CC2935 CC2938 CC3059 CC3061 CC3062 CC3063 CC3073 CC3075 CC3079 CC3084\".split()\n",
    "\n",
    "\n",
    "#### STATS TO COLLECT for each gene\n",
    "genes = []\n",
    "mtPlus_coords = []\n",
    "mtMinus_coords = []\n",
    "alleles_ALL = []\n",
    "sites_ALL = []\n",
    "invariant_ALL = []\n",
    "variant_ALL = []\n",
    "theta_pi_total_ALL = []\n",
    "theta_w_total_ALL = []\n",
    "sfs_t_ALL = []\n",
    "# Loop over all shared coding gametologs\n",
    "alignment_files = glob.glob(\"../../../data/aligned-fastas/*non_gametolog.fasta\")\n",
    "print(alignment_files)\n",
    "\n",
    "for a in alignment_files[:]:\n",
    "    !head $a -n 1 \n",
    "#     print(a)\n",
    "#     gene = a.split(\"/\")[-1].split(\".\")[0]\n",
    "#     mt_allele = a.split(\"/\")[-1].split(\".\")[-2].split(\"_\")[-1]\n",
    "#     print(gene, mt_allele)\n",
    "#     genes.append(gene)\n",
    "#     aln = AlignIO.read(open(a), 'fasta')\n",
    "#     if mt_allele == \"mtPlus\":\n",
    "#         mtPlus_sampleIDs  = [s.id for s in aln if s.id.split(\"|\")[0] in mtPlus_Genotypes]\n",
    "#         mtPlus_coords.append(mtPlus_sampleIDs[0].split(\"|\")[-1]) ## based on this kind of name e.g. \"CC2935|mtMinus:105065-114642\"\n",
    "#         mtMinus_coords.append('N/A')\n",
    "#     else:\n",
    "#         mtMinus_sampleIDs = [s.id for s in aln if s.id.split(\"|\")[0] in mtMinus_Genotypes]\n",
    "#         mtMinus_coords.append(mtMinus_sampleIDs[0].split(\"|\")[-1])\n",
    "#         mtPlus_coords.append(\"N/A\")\n",
    "#     degen_map = degenerate(special_consensus(aln))\n",
    "#     sfst,sfs0,sfs4 = alignment2sfs_w_degen(aln, 6, degen_map, samples=None, start=None, end=None)\n",
    "#     ##### record ALL results\n",
    "#     alleles_ALL.append(sfst.alleles)\n",
    "#     sites_ALL.append(sfst.sites())\n",
    "#     invariant_ALL.append(sfst.invariant())\n",
    "#     variant_ALL.append(sfst.variant())\n",
    "#     theta_pi_0fold_ALL.append(sfs0.theta_pi())\n",
    "#     theta_pi_4fold_ALL.append(sfs4.theta_pi())\n",
    "#     theta_pi_total_ALL.append(sfst.theta_pi())\n",
    "#     theta_w_0fold_ALL.append(sfs0.theta_w())\n",
    "#     theta_w_4fold_ALL.append(sfs4.theta_w())\n",
    "#     theta_w_total_ALL.append(sfst.theta_w())\n",
    "#     sfs_4fold_ALL.append(sfs4.sfs)\n",
    "#     sfs_0fold_ALL.append(sfs0.sfs)\n",
    "#     sfs_t_ALL.append(sfst.sfs)\n",
    "#     #######\n",
    "#     ## GC_4fold\n",
    "#      GC4.append(len([1 for i,j in zip(special_consensus(aln),degen_map) if i in 'GCgc' and j =='4'])/\n",
    "#                 len([1 for i,j in zip(special_consensus(aln),degen_map) if i in 'AaTtGCgc' and j =='4']))\n",
    "   \n",
    "# #########################################################\n",
    "# #### DF the results #####################################\n",
    "# #########################################################\n",
    "# mtLimited_results = pd.DataFrame(\n",
    "#     {'gene' : genes,\n",
    "#      'mtPlus_coords' : mtPlus_coords,\n",
    "#      'mtMinus_coords' : mtMinus_coords,\n",
    "#      'alleles_ALL' : alleles_ALL, \n",
    "#      'sites_ALL' : sites_ALL, \n",
    "#      'invariant_ALL' : invariant_ALL, \n",
    "#      'variant_ALL' : variant_ALL, \n",
    "#      'theta_pi_0fold_ALL' : theta_pi_0fold_ALL, \n",
    "#      'theta_pi_4fold_ALL' : theta_pi_4fold_ALL, \n",
    "#      'theta_pi_total_ALL' : theta_pi_total_ALL, \n",
    "#      'theta_w_0fold_ALL' : theta_w_0fold_ALL, \n",
    "#      'theta_w_4fold_ALL' : theta_w_4fold_ALL, \n",
    "#      'theta_w_total_ALL' : theta_w_total_ALL, \n",
    "#      'sfs_4fold_ALL' : sfs_4fold_ALL, \n",
    "#      'sfs_0fold_ALL' : sfs_0fold_ALL, \n",
    "#      'sfs_t_ALL' : sfs_t_ALL,    \n",
    "#      'GC4' : GC4\n",
    "#     })  \n",
    "# mtLimited_results.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Windowed Gametolog #AHMED\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "  0%|          | 0/646 [00:00<?, ?it/s]\u001b[A\n",
      "  0%|          | 1/646 [00:40<7:15:29, 40.51s/it]\n",
      " 12%|        | 77/646 [51:38<6:23:46, 40.47s/it]"
     ]
    },
    {
     "ename": "ZeroDivisionError",
     "evalue": "division by zero",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mZeroDivisionError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-20-4cb078644f25>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    180\u001b[0m     \u001b[0;31m###################################\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    181\u001b[0m     \u001b[0;31m## population stats\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 182\u001b[0;31m     \u001b[0mpop_stats\u001b[0m\u001b[0;34m=\u001b[0m \u001b[0mpopulation_poly\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maln\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m12\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdegen_map\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmtPlus_sampleIDs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmtMinus_sampleIDs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstart\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maln_start\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maln_end\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    183\u001b[0m     \u001b[0mprivate_MTPLUS\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m  \u001b[0mpop_stats\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'private1'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    184\u001b[0m     \u001b[0mprivate_MTMINUS\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m \u001b[0mpop_stats\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'private2'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-3-6834364177d1>\u001b[0m in \u001b[0;36mpopulation_poly\u001b[0;34m(aln, min_called, degen_map, group1, group2, start, end)\u001b[0m\n\u001b[1;32m    158\u001b[0m                 \u001b[0;34m'fixed'\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mfixed\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    159\u001b[0m                 \u001b[0;34m'shared'\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mshared\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 160\u001b[0;31m                 'fst' :fst[0]/fst[1]}\n\u001b[0m\u001b[1;32m    161\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mresults\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    162\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mZeroDivisionError\u001b[0m: division by zero"
     ]
    }
   ],
   "source": [
    "from tqdm import tqdm\n",
    "# Target Genotype Lists\n",
    "mtPlus_Genotypes  = \"CC2936 CC2937 CC3060 CC3064 CC3065 CC3068 CC3071 CC3076 CC3086\".split()\n",
    "mtMinus_Genotypes = \"CC2935 CC2938 CC3059 CC3061 CC3062 CC3063 CC3073 CC3075 CC3079 CC3084\".split()\n",
    "\n",
    "\n",
    "#### STATS TO COLLECT \n",
    "windows = []\n",
    "start_coords = []\n",
    "end_coords = []\n",
    "alleles_ALL = []\n",
    "sites_ALL = []\n",
    "invariant_ALL = []\n",
    "variant_ALL = []\n",
    "theta_pi_notSilent_ALL = []\n",
    "theta_pi_silent_ALL = []\n",
    "theta_pi_total_ALL = []\n",
    "theta_w_notSilent_ALL = []\n",
    "theta_w_silent_ALL = []\n",
    "theta_w_total_ALL = []\n",
    "sfs_silent_ALL = []\n",
    "sfs_notSilent_ALL = []\n",
    "sfs_t_ALL = []\n",
    "alleles_MTPLUS = []\n",
    "sites_MTPLUS = []\n",
    "invariant_MTPLUS = []\n",
    "variant_MTPLUS = []\n",
    "theta_pi_notSilent_MTPLUS = []\n",
    "theta_pi_silent_MTPLUS = []\n",
    "theta_pi_total_MTPLUS = []\n",
    "theta_w_notSilent_MTPLUS = []\n",
    "theta_w_silent_MTPLUS = []\n",
    "theta_w_total_MTPLUS = []\n",
    "sfs_silent_MTPLUS = []\n",
    "sfs_notSilent_MTPLUS = []\n",
    "sfs_t_MTPLUS = []\n",
    "alleles_MTMINUS = []\n",
    "sites_MTMINUS = []\n",
    "invariant_MTMINUS = []\n",
    "variant_MTMINUS = []\n",
    "theta_pi_notSilent_MTMINUS = []\n",
    "theta_pi_silent_MTMINUS = []\n",
    "theta_pi_total_MTMINUS = []\n",
    "theta_w_notSilent_MTMINUS = []\n",
    "theta_w_silent_MTMINUS = []\n",
    "theta_w_total_MTMINUS = []\n",
    "sfs_silent_MTMINUS = []\n",
    "sfs_notSilent_MTMINUS = []\n",
    "sfs_t_MTMINUS = []\n",
    "private_MTPLUS = []\n",
    "private_MTMINUS = []\n",
    "shared = []\n",
    "fixed = []\n",
    "Fst = []\n",
    "GCsilent = []\n",
    "\n",
    "\n",
    "# To calculate silent/vs non-silent stats I can manipulate the degen map as follows:\n",
    "# - 4 = all silent sites -intron, intergenic, 4-fold\n",
    "# I need to use the annotation table for the region   \n",
    "# 298298-943474 (0-based python style)\n",
    "\n",
    "import annotation_table, pysam\n",
    "annotation_table_file = \"/scratch/research/references/chlamydomonas/5.3_chlamy_w_organelles_mt_minus/annotation/concatenated_GFF/annotation_table.txt.gz\"\n",
    "at = pysam.Tabixfile(annotation_table_file)\n",
    "special_degen_map = ''\n",
    "for l in  at.fetch('chromosome_6',298298,943474): #note used 0-based non inclusive coords\n",
    "    a = annotation_table.annotation_line(l)\n",
    "    if a.fold4 =='1' or a.intergenic =='1' or a.intronic =='1': # all silent sites are marked as 4-fold\n",
    "        special_degen_map+='4'\n",
    "    else:\n",
    "        special_degen_map+='0'\n",
    "\n",
    "a = (\"../../../data/aligned-fastas/mt_aligned_all.fasta\")\n",
    "aln = AlignIO.read(open(a), 'fasta')\n",
    "\n",
    "#Ahmed Windows\n",
    "# Add windows\n",
    "window_coords = [] # this should be [start1, end1], [start2, end2]... in python coords\n",
    "\n",
    "def get_genomic_coords(start, end, windowsize, offset):\n",
    "    ''' \n",
    "    returns genomic start and end of entire region,\n",
    "    rounded to match windowsize\n",
    "    start: alignment python index start (0)\n",
    "    end: alignment python index end\n",
    "    windowsize: desired windowsize\n",
    "    offset: what genomic position is at alignment index 0?\n",
    "    '''\n",
    "    genomic_start = round((start + offset) / windowsize) * windowsize\n",
    "    genomic_end = (round((end + offset) / windowsize) * windowsize) + windowsize\n",
    "    if genomic_start > offset:\n",
    "        genomic_start = genomic_start - windowsize\n",
    "    if genomic_end - windowsize > offset + end:\n",
    "        genomic_end = genomic_end - windowsize\n",
    "    return genomic_start, genomic_end\n",
    "\n",
    "def get_window_coords(start, end, windowsize, offset):\n",
    "    ''' \n",
    "    returns python index windows\n",
    "    start: alignment python index start (ie 0)\n",
    "    end: alignment python index end\n",
    "    windowsize: desired windowsize\n",
    "    offset: what genomic position is at alignment index 0?\n",
    "    '''\n",
    "    genomic_start, genomic_end = get_genomic_coords(start, end, windowsize, offset)\n",
    "    genomic_windows = list(range(genomic_start, genomic_end + windowsize, windowsize))\n",
    "    aln_offset = windowsize - (offset - genomic_windows[0])\n",
    "    window_coords_all = [0] + [aln_offset + (windowsize * i) for i in range(len(genomic_windows) - 1)]\n",
    "    window_coords_all[-1] = end # avoid IndexError in final window\n",
    "    window_coords_nested = [window_coords_all[i:i+2] for i in range(len(window_coords_all) - 1)]\n",
    "    return window_coords_nested\n",
    "    \n",
    "start = 0\n",
    "end = 645176\n",
    "windowsize = 1000 # to match zns file\n",
    "offset = 298298\n",
    "windows_coords = get_window_coords(start, end, windowsize, offset)\n",
    "\n",
    "for w in tqdm(windows_coords):\n",
    "    aln_start, aln_end = w\n",
    "    window = \"%i-%i\" %(aln_start, aln_end)\n",
    "    windows.append(window)\n",
    "    mtPlus_sampleIDs  = [s.id for s in aln if s.id.split(\"|\")[0] in mtPlus_Genotypes]\n",
    "    mtMinus_sampleIDs = [s.id for s in aln if s.id.split(\"|\")[0] in mtMinus_Genotypes]\n",
    "    genome_start = aln_start + offset # globally set offset is 298298\n",
    "    genome_end = aln_end + offset\n",
    "    start_coords.append(genome_start) \n",
    "    end_coords.append(genome_end)\n",
    "    degen_map = special_degen_map\n",
    "    sfst,sfs0,sfs4 = alignment2sfs_w_degen(aln, 12, degen_map, samples=None, start=aln_start, end=aln_end)\n",
    "    ##### record ALL results\n",
    "    alleles_ALL.append(sfst.alleles)\n",
    "    sites_ALL.append(sfst.sites())\n",
    "    invariant_ALL.append(sfst.invariant())\n",
    "    variant_ALL.append(sfst.variant())\n",
    "    theta_pi_notSilent_ALL.append(sfs0.theta_pi())\n",
    "    theta_pi_silent_ALL.append(sfs4.theta_pi())\n",
    "    theta_pi_total_ALL.append(sfst.theta_pi())\n",
    "    theta_w_notSilent_ALL.append(sfs0.theta_w())\n",
    "    theta_w_silent_ALL.append(sfs4.theta_w())\n",
    "    theta_w_total_ALL.append(sfst.theta_w())\n",
    "    sfs_silent_ALL.append(sfs4.sfs)\n",
    "    sfs_notSilent_ALL.append(sfs0.sfs)\n",
    "    sfs_t_ALL.append(sfst.sfs)\n",
    "    ###################################\n",
    "    ## MTPLUS\n",
    "    sfst,sfs0,sfs4 = alignment2sfs_w_degen(aln, 8, degen_map, samples=mtPlus_sampleIDs, start=aln_start, end=aln_end)\n",
    "    ##### record ALL results\n",
    "    alleles_MTPLUS.append(sfst.alleles)\n",
    "    sites_MTPLUS.append(sfst.sites())\n",
    "    invariant_MTPLUS.append(sfst.invariant())\n",
    "    variant_MTPLUS.append(sfst.variant())\n",
    "    theta_pi_notSilent_MTPLUS.append(sfs0.theta_pi())\n",
    "    theta_pi_silent_MTPLUS.append(sfs4.theta_pi())\n",
    "    theta_pi_total_MTPLUS.append(sfst.theta_pi())\n",
    "    theta_w_notSilent_MTPLUS.append(sfs0.theta_w())\n",
    "    theta_w_silent_MTPLUS.append(sfs4.theta_w())\n",
    "    theta_w_total_MTPLUS.append(sfst.theta_w())\n",
    "    sfs_silent_MTPLUS.append(sfs4.sfs)\n",
    "    sfs_notSilent_MTPLUS.append(sfs0.sfs)\n",
    "    sfs_t_MTPLUS.append(sfst.sfs)   \n",
    "    ###################################\n",
    "    ## MTMINUS\n",
    "    sfst,sfs0,sfs4 = alignment2sfs_w_degen(aln, 8, degen_map, samples=mtMinus_sampleIDs, start=aln_start, end=aln_end)\n",
    "    ##### record ALL results\n",
    "    alleles_MTMINUS.append(sfst.alleles)\n",
    "    sites_MTMINUS.append(sfst.sites())\n",
    "    invariant_MTMINUS.append(sfst.invariant())\n",
    "    variant_MTMINUS.append(sfst.variant())\n",
    "    theta_pi_notSilent_MTMINUS.append(sfs0.theta_pi())\n",
    "    theta_pi_silent_MTMINUS.append(sfs4.theta_pi())\n",
    "    theta_pi_total_MTMINUS.append(sfst.theta_pi())\n",
    "    theta_w_notSilent_MTMINUS.append(sfs0.theta_w())\n",
    "    theta_w_silent_MTMINUS.append(sfs4.theta_w())\n",
    "    theta_w_total_MTMINUS.append(sfst.theta_w())\n",
    "    sfs_silent_MTMINUS.append(sfs4.sfs)\n",
    "    sfs_notSilent_MTMINUS.append(sfs0.sfs)\n",
    "    sfs_t_MTMINUS.append(sfst.sfs)   \n",
    "    ###################################\n",
    "    ## population stats\n",
    "    pop_stats= population_poly(aln, 12, degen_map, mtPlus_sampleIDs, mtMinus_sampleIDs, start=aln_start, end=aln_end)\n",
    "    private_MTPLUS.append(  pop_stats['private1'])\n",
    "    private_MTMINUS.append( pop_stats['private2'])\n",
    "    shared.append(pop_stats['shared'])\n",
    "    fixed.append(pop_stats['fixed'])\n",
    "    Fst.append(pop_stats['fst'])\n",
    "    #######\n",
    "    ## GC_silent\n",
    "    GCsilent.append(len([1 for i,j in zip(special_consensus(aln)[aln_start:aln_end],degen_map[aln_start:aln_end]) if i in 'GCgc' and j =='4'])/\n",
    "               len([1 for i,j in      zip(special_consensus(aln)[aln_start:aln_end],degen_map[aln_start:aln_end]) if i in 'AaTtGCgc' and j =='4']))\n",
    "\n",
    "\n",
    "#########################################################\n",
    "#### DF the results #####################################\n",
    "#########################################################\n",
    "print(GCsilent)\n",
    "gametolog_results = pd.DataFrame(\n",
    "    {'window' : windows,\n",
    "     'start_coords' : start_coords,\n",
    "     'end_coords' : end_coords,\n",
    "     'alleles_ALL' : alleles_ALL, \n",
    "     'sites_ALL' : sites_ALL, \n",
    "     'invariant_ALL' : invariant_ALL, \n",
    "     'variant_ALL' : variant_ALL, \n",
    "     'theta_pi_notSilent_ALL' : theta_pi_notSilent_ALL, \n",
    "     'theta_pi_silent_ALL' : theta_pi_silent_ALL, \n",
    "     'theta_pi_total_ALL' : theta_pi_total_ALL, \n",
    "     'theta_w_notSilent_ALL' : theta_w_notSilent_ALL, \n",
    "     'theta_w_silent_ALL' : theta_w_silent_ALL, \n",
    "     'theta_w_total_ALL' : theta_w_total_ALL, \n",
    "     'sfs_silent_ALL' : sfs_silent_ALL, \n",
    "     'sfs_notSilent_ALL' : sfs_notSilent_ALL, \n",
    "     'sfs_t_ALL' : sfs_t_ALL,    \n",
    "     'alleles_MTPLUS' : alleles_MTPLUS, \n",
    "     'sites_MTPLUS' : sites_MTPLUS, \n",
    "     'invariant_MTPLUS' : invariant_MTPLUS, \n",
    "     'variant_MTPLUS' : variant_MTPLUS, \n",
    "     'theta_pi_notSilent_MTPLUS' : theta_pi_notSilent_MTPLUS, \n",
    "     'theta_pi_silent_MTPLUS' : theta_pi_silent_MTPLUS, \n",
    "     'theta_pi_total_MTPLUS' : theta_pi_total_MTPLUS, \n",
    "     'theta_w_notSilent_MTPLUS' : theta_w_notSilent_MTPLUS, \n",
    "     'theta_w_silent_MTPLUS' : theta_w_silent_MTPLUS, \n",
    "     'theta_w_total_MTPLUS' : theta_w_total_MTPLUS, \n",
    "     'sfs_silent_MTPLUS' : sfs_silent_MTPLUS, \n",
    "     'sfs_notSilent_MTPLUS' : sfs_notSilent_MTPLUS, \n",
    "     'sfs_t_MTPLUS' : sfs_t_MTPLUS,         \n",
    "     'alleles_MTMINUS' : alleles_MTMINUS, \n",
    "     'sites_MTMINUS' : sites_MTMINUS, \n",
    "     'invariant_MTMINUS' : invariant_MTMINUS, \n",
    "     'variant_MTMINUS' : variant_MTMINUS, \n",
    "     'theta_pi_notSilent_MTMINUS' : theta_pi_notSilent_MTMINUS, \n",
    "     'theta_pi_silent_MTMINUS' : theta_pi_silent_MTMINUS, \n",
    "     'theta_pi_total_MTMINUS' : theta_pi_total_MTMINUS, \n",
    "     'theta_w_notSilent_MTMINUS' : theta_w_notSilent_MTMINUS, \n",
    "     'theta_w_silent_MTMINUS' : theta_w_silent_MTMINUS, \n",
    "     'theta_w_total_MTMINUS' : theta_w_total_MTMINUS, \n",
    "     'sfs_silent_MTMINUS' : sfs_silent_MTMINUS, \n",
    "     'sfs_notSilent_MTMINUS' : sfs_notSilent_MTMINUS, \n",
    "     'sfs_t_MTMINUS' : sfs_t_MTMINUS, \n",
    "     'private_MTPLUS' : private_MTPLUS, \n",
    "     'private_MTMINUS' : private_MTMINUS, \n",
    "     'shared' : shared, \n",
    "     'fixed' : fixed, \n",
    "     'Fst' : Fst,\n",
    "     'GCsilent' : GCsilent\n",
    "    })\n",
    "\n",
    "\n",
    "gametolog_results.to_csv(\"polymorphism_all_windowed.gametolog.csv\", index=False)\n",
    "gametolog_results.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "jupytext": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
