---
title: "MT locus recombination plots"
author: Ahmed Hasan
output: pdf_document
---

```{r, warning=FALSE}
library(readr)
library(dplyr, warn.conflicts = FALSE)
library(magrittr)
library(tidyr, warn.conflicts = FALSE)
library(ggplot2)
library(purrr, warn.conflicts = FALSE)
library(patchwork)
library(wesanderson)
library(stringr)
library(fs)
library(depmixS4)
library(viridis)
library(ggalt) # for relative annotations
library(broom)
library(car, warn.conflicts = FALSE)
library(emmeans, warn.conflicts = FALSE)
library(ggridges, warn.conflicts = FALSE)
library(ggpubr)
select <- dplyr::select

setwd('~/Desktop/Coding/GitHub/2019-mt-locus/')

options(contrasts = c("contr.sum", "contr.poly"))
# options(contrasts = c("contr.treatment", "contr.poly")) # defaults
```


# Main Text

## Figure 1 - ZnS across the MT locus

Data preparation:

```{r}
fnames <- fs::dir_ls('data/ld-windowed', glob = '*zns*txt')
zns_files <- map_dfr(fnames, read_delim, delim = ' ', col_types = cols(), .id = 'name') %>% 
  mutate(region = str_extract(name, '(mt|plus|chromosome_6)')) %>% 
  mutate(midpoint = start + 500) %>% 
  filter(site_count != 0) %>% 
  dplyr::select(-name)

zns_plot_theme <- function(font_size = 16) {
  theme(axis.title = element_text(family = "Helvetica", size = font_size),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        axis.text = element_text(family = "Helvetica", size = font_size, color = 'black'),
        axis.line = element_line(colour = 'black', linetype = 'solid'),
        axis.line.x = element_line(size = 0.9),
        axis.line.y = element_line(size = 0.9),
        panel.background = element_blank())
}

mt_only <- zns_files %>% 
  filter(region == 'mt') %>% 
  filter(site_count > 30) %>% 
  mutate(
    domain = case_when(
      start <= 420000 ~ 'T',
      start > 420000 & start <= 826000 ~ 'R',
      start > 826000 ~ 'C'
    )
  )

chr6_only <- zns_files %>% 
  filter(site_count > 30) %>% 
  filter(region == 'chromosome_6') %>% 
  filter(start < 298000 | end > 943000)

plot_data <- bind_rows(mt_only, chr6_only) %>% 
  arrange(start)

```

HMM fit:

```{r}
hmm_fit <- function(d, nstates) {
  model <- d %>% 
    depmix(
      list(zns ~ 1), data = .,
      family = list(gaussian()), nstates = nstates)
  model_fit <- fit(model)
  posterior_states <- posterior(model_fit)
  return(posterior_states)
}

posterior_3 <- hmm_fit(plot_data, 3)

# add HMM fit to plot_data
plot_data <- bind_cols(plot_data, dplyr::select(posterior_3, state))
plot_data$state <- as.factor(plot_data$state)
```

ZnS across the first ~3 Mb of chromosome 6:

```{r}
hmm_colors <- viridis(3)
zns_plot_all <- plot_data %>% # only gametologs and chr6
  ggplot(aes(x = midpoint, y = zns, color = state)) +
  geom_rect(aes(xmin = 298000, xmax = 420000, ymin = -0.5, ymax = 1.5),  # 0-175k in bed - previously 473000
            color = '#eaecef', fill = '#eaecef') +
  geom_rect(aes(xmin = 421000, xmax = 826000, ymin = -0.5, ymax = 1.5),
            color = '#d9dce0', fill = '#d9dce0') +
  geom_rect(aes(xmin = 827000, xmax = 943000, ymin = -0.5, ymax = 1.5), 
            color = '#eaecef', fill = '#eaecef') +
  geom_point(size = 1.2) +
  labs(
    x = 'position on chromosome 6 (bp)',
    y = expression(paste(Z[nS]))
  ) +
  scale_color_manual(
    values = c(hmm_colors[1], hmm_colors[2], hmm_colors[3])) +
  scale_x_continuous(
    breaks = seq(0, 40, by = 5)*1e5, 
    labels = scales::comma,
    limits = c(0, 31.5)*1e5) +
  coord_cartesian(
    x = c(0, 30.5)*1e5, y = c(0, 1)) +
  zns_plot_theme(16) +
  annotate('text', label = 'T', x = 360000, y = 0.05, size = 8) +
  annotate('text', label = 'R', x = 638000, y = 0.05, size = 8) +
  annotate('text', label = 'C', x = 884000, y = 0.05, size = 8) +
  guides(color = FALSE) +
  NULL

zns_plot_all

ggsave('plots/fig_1.eps', plot = zns_plot_all,
       width = par('din')[1] * 1.75, height = par('din')[1])

```


## Figure 2 - Fst and ZnS over shared regions

Data preparation:

```{r}
set.seed(42)
polymorphism_shared <- read_csv('data/cds-popgen/polymorphism_all.shared.20190225.zns.csv', col_types = cols()) %>% 
  mutate(piN_piS = theta_pi_0fold_ALL / theta_pi_4fold_ALL) %>% 
  mutate(piN_piS_PLUS = theta_pi_0fold_MTPLUS / theta_pi_4fold_MTPLUS) %>% 
  mutate(piN_piS_MINUS = theta_pi_0fold_MTMINUS / theta_pi_4fold_MTMINUS) %>% 
  mutate(chrom = 'shared') # for boxplot later
polymorphism_limited <- read_csv('data/cds-popgen/polymorphism.mtLimited.zns.csv', col_types = cols()) %>% 
  mutate(piN_piS = theta_pi_0fold_ALL / theta_pi_4fold_ALL) %>% 
  mutate(chrom = 'limited')
polymorphism_autosomal <- read_csv('data/cds-popgen/polymorphism.nonMT.csv', col_types = cols()) %>% 
  separate(mtPlus_coords, into = c('chrom', 'pos'), sep = ':') %>% 
  mutate(piN_piS = theta_pi_0fold_ALL / theta_pi_4fold_ALL)
polymorphism_chr6 <- polymorphism_autosomal %>% 
  filter(chrom == 'chromosome_6') %>% 
  sample_n(100)
polymorphism_other <- polymorphism_autosomal %>% 
  filter(chrom != 'chromosome_6') %>% 
  mutate(chrom = 'other') %>% 
  sample_n(100)
windowed <- read_csv(
  'data/cds-popgen/polymorphism_all_windowed_script.gametolog.csv',
  col_types = cols()
  ) %>% 
  rename(start = start_coords)
fst_chr6 <- read_csv(
  'data/cds-popgen/polymorphism_fst_windowed.chr6.csv', 
  col_types = cols()) %>% 
  select(start = start_coords, end = end_coords, everything()) %>% 
  left_join(
    read_delim(
      'data/ld-windowed/autosomal/chromosome_6_final.zns',
      delim = ' ', col_types = cols()),
    by = c('start', 'end')
  ) %>% 
  mutate(Fst = ifelse(Fst == 'N/A', 0, as.numeric(Fst))) %>% 
  filter(start < 298000 | start > 943000) %>% 
  filter(site_count >= 30) %>% 
  mutate(Fst = ifelse(Fst < 0, 0, Fst))

fst_chr6 %>% 
  lm(Fst ~ zns, data = .) %>% 
  summary()

ggplot(fst_chr6, aes(x = zns, y = Fst)) +
  geom_point(alpha = 0.4) +
  geom_smooth(method = 'lm') +
  theme_classic()

windowed$start[1] <- 298000 # to match zns file

windowed_all <- left_join(
  windowed, filter(zns_files, region == 'mt'), by = 'start'
  ) %>% 
  mutate(
    Fst = as.numeric(Fst), 
    GCsilent = as.numeric(GCsilent)
  )

windowed_filtered <- windowed_all %>% 
  filter(!is.na(Fst), !is.na(zns)) %>% 
  filter(site_count >= 30) %>% 
  select(Fst, zns, start, end, midpoint, region) %>% 
  mutate(
    domain = case_when(
      start <= 420000 ~ 'T',
      start > 420000 & start <= 826000 ~ 'R',
      start > 826000 ~ 'C'
    )
  )

fst_chr6 %>% 
  filter(start < 298000 | start > 943000) %>% 
  filter(site_count > 30) %>% 
  mutate(window = floor(start / 10000) * 10000) %>% 
  group_by(window) %>% 
  summarise(Fst = mean(Fst, na.rm = TRUE),
            zns = mean(zns, na.rm = TRUE)) %>% 
  ggplot(., aes(x = zns, y = Fst)) +
  geom_point() +
  geom_smooth(method = 'lm')

```

Fst and ZnS across the three domains:

```{r}
corr_plot_theme <- function(font_size = 16) {
  theme(axis.title = element_text(family = "Helvetica", size = font_size),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        axis.text = element_text(family = "Helvetica", size = font_size, color = 'black'),
        axis.line = element_line(colour = 'black', linetype = 'solid'),
        axis.line.x = element_line(size = 0.9),
        axis.line.y = element_line(size = 0.9),
        axis.ticks = element_line(colour = 'black'),
        plot.tag = element_text(colour = 'black', size = font_size, face = 'bold'),
        panel.background = element_blank(),
        strip.background = element_blank(),
        strip.text = element_text(family = 'Helvetica', face = 'bold', colour = 'black', size = font_size),
        strip.placement = 'inside')
}
  

fst_plot_domains <- windowed_filtered %>% 
  mutate(domain_f = factor(domain, levels = c('T', 'R', 'C'))) %>% 
  ggplot(aes(x = zns, y = Fst, color = domain_f)) +
  geom_point(aes(color = domain_f), size = 1.5, alpha = 1) +
  geom_smooth(method = 'lm', se = FALSE, size = 0.5) +
  labs(
    x = expression(Z[nS]),
    y = expression(F[ST]),
    color = 'Domain'
  ) +
  coord_cartesian(
    x = c(0, 1.3),
    y = c(0, 1)
  ) +
  ylim(0, 1) +
  scale_x_continuous(breaks = c(0, 0.25, 0.5, 0.75, 1)) +
  corr_plot_theme(16) +
  theme(legend.key = element_blank(),
        legend.title = element_text(family = 'Helvetica', size = 16),
        legend.text = element_text(family = 'Helvetica', size = 14),
        legend.position = c(0.1, 0.8)) +
  scale_color_manual(
    values = c(
      'T' = wes_palette('Darjeeling1')[2],
      'R' = wes_palette('Darjeeling1')[3],
      'C' = wes_palette('Darjeeling1')[4]
    )
  ) +
  annotate('text', x = 1.175, y = 0.95, size = 5,
           label = 'italic(R) ^ 2 == "0.799***"',
           parse = TRUE) +
  annotate('text', x = 1.12, y = 0.625, size = 5,
           label = 'italic(R) ^ 2 == "0.389***"',
           parse = TRUE) +
  annotate('text', x = 1.1, y = 0.215, size = 5,
           label = 'italic(R) ^ 2 == "0.0005"',
           parse = TRUE)

fst_plot_domains

ggsave('plots/fig_2.eps', plot = fst_plot_domains,
       width = par('din')[1] * 1.25, height = par('din')[1] * 0.75)
```

Stats:

```{r}
# multiple regression
zns_fit <- lm(Fst ~ zns * domain, data = windowed_filtered)
summary(zns_fit) # r2 = 0.8462, p < 2.2e-16

# interaction
zns_anova <- car::Anova(zns_fit, type = 3) # interaction - p = 2.4e-08

# domain-specific regressions
windowed_filtered %>% 
  split(.$domain) %>% 
  map_dfr(~ lm(Fst ~ zns, data = .) %>% 
            summary() %>% 
            tidy(),
          .id = 'name')

```


## Figure 3 - pi and diversity

Data preparation in Figure 2 section above

### Fig. 3A - pi and ZnS over both mating types

MT-separated regressions:

```{r}
lm(theta_pi_4fold_MTPLUS ~ zns_all, data = polymorphism_shared) %>% 
  summary() # R2 = 0.1395, p = 0.016
  
lm(theta_pi_4fold_MTMINUS ~ zns_all, data = polymorphism_shared) %>% 
  summary() # R2 = 0.187, p = 0.0047
```

Silent pi in all, MT+, and MT- in shared regions:

```{r}
polymorphism_other %>% # non chr6 genes
  summarise_at(vars(starts_with('theta_pi_4fold')), mean, na.rm = TRUE) %>% 
  gather(measure, value)

polymorphism_chr6 %>% # chr6 genes
  summarise_at(vars(starts_with('theta_pi_4fold')), mean, na.rm = TRUE) %>% 
  gather(measure, value)

polymorphism_shared %>% 
  summarise_at(vars(starts_with('theta_pi_4fold')), mean, na.rm = TRUE) %>% 
  gather(measure, value) # all three are lower than autosomal levels, but MT+ especially low
```

Silent pi in MT+ and MT- limited regions:

```{r}
polymorphism_limited %>% 
  mutate(allele = case_when(
    !is.na(mtPlus_coords) ~ 'plus',
    !is.na(mtMinus_coords) ~ 'minus')
  ) %>% 
  group_by(allele) %>% 
  summarise(mean_theta = mean(theta_pi_4fold_ALL))

```


Plot:

```{r}
pi_mt_plot <- ggplot(polymorphism_shared, aes(x = zns_all)) +
  geom_point(
    data = polymorphism_shared,
    aes(y = theta_pi_4fold_MTPLUS),
    size = 2,
    color = wes_palette('Darjeeling1')[2]
  ) + # plus
  geom_point(
    data = polymorphism_shared,
    aes(y = theta_pi_4fold_MTMINUS), 
    size = 2,
    color = wes_palette('Darjeeling1')[3]) +
  geom_smooth(
    data = polymorphism_shared,
    aes(y = theta_pi_4fold_MTPLUS), 
    method = 'lm',
    se = FALSE,
    color = wes_palette('Darjeeling1')[2]) + # plus
  geom_smooth(
    data = polymorphism_shared,
    aes(y = theta_pi_4fold_MTMINUS), method = 'lm',
    se = FALSE,
    color = wes_palette('Darjeeling1')[3]) +
  corr_plot_theme(16) +
  labs(
    x = expression(paste(Z[nS])),
    y = expression(paste(pi[silent])),
    tags = 'a'
  ) +
  coord_cartesian(
    x = c(0, 1),
    y = c(0, 0.045)
  ) +
  annotate(
    'text', x = 0.9, y = 0.04, 
    size = 6, label = 'MT+',
    color = wes_palette('Darjeeling1')[2]) +
  annotate(
    'text', x = 0.9, y = 0.035, 
    size = 6, label = 'MT-',
    color = wes_palette('Darjeeling1')[3]) +
  NULL
  
pi_mt_plot
```

### Fig. 3B - piN/piS and ZnS

Data preparation + regression:

```{r}
pi_all <- list(polymorphism_shared, polymorphism_limited) %>% 
  map_dfr(~ select(., gene, contains('piN_piS'), sites_ALL, zns_all), .id = 'name') %>% 
  mutate(
    is_gametolog = case_when(
      name == 1 ~ 1,
      name == 2 ~ 0
    )
  ) %>% 
  select(-name)

pi_all %>% 
  filter(is_gametolog == 1) %>% 
  filter(gene != '522915', gene != 'MT0796') %>% 
  select(piN_piS, zns_all, sites_ALL) %>% 
  lm(piN_piS ~ zns_all, data = ., weights = sites_ALL) %>% 
  summary() # R2 = 0.106, p = 0.043

# incl. weird genes
pi_all %>% 
  filter(is_gametolog == 1) %>%
  select(piN_piS, zns_all, sites_ALL) %>% 
  lm(piN_piS ~ zns_all, data = ., weights = sites_ALL) %>% 
  summary() # R2 = 0.094, p = 0.051
```

Plot:

```{r}
pi_zns_plot <- polymorphism_shared %>% 
  filter(gene != 'MT0796', gene != '522915') %>% 
  ggplot(aes(x = zns_all, y = piN_piS)) +
  geom_point(size = 2) +
  geom_smooth(method = 'lm', se = FALSE) +
  corr_plot_theme(16) +
  labs(
    x = expression(paste(Z[nS])),
    y = expression(paste(pi[N]/pi[S])),
    tag = 'b'
  ) +
  coord_cartesian(
    x = c(0, 1),
    y = c(0, 1)
  ) +
  annotate(
    'text', x = 0.2, y = 0.9, size = 6, 
     label = 'italic(R) ^ 2 == 0.106',
     parse = TRUE) +
  annotate(
    'text', x = 0.2, y = 0.8, size = 6,
    label = 'p == 0.043', parse = TRUE)

pi_zns_plot

ggsave('plots/piNpiS_zns_genes.pdf', plot = pi_zns_plot)
```

### Combining Fig. 3A and 3B

```{r}
fig_3 <- pi_mt_plot + pi_zns_plot
fig_3

ggsave('plots/fig_3.eps', plot = fig_3,
       width = par('din')[1] * 1.5, height = par('din')[1] * 1.5 * 0.5)
```

## Table 1 - piN/piS and GC content stats over different annotations

### piN/piS

```{r}
variation_all <- lapply(ls(pattern = 'polymorphism*'), get)
names(variation_all) <- ls(pattern = 'polymorphism*')

variation_all %>% 
  map_dfr(~ filter(., is.finite(piN_piS), !is.na(piN_piS)) %>% 
  summarise(mean_piN_piS = mean(piN_piS),
            median_piN_piS = median(piN_piS)), .id = 'name')
```

### GC content

```{r}
variation_all %>% 
  map_dfr(~ filter(., !is.na(GC4)) %>% 
  summarise(mean_GC4 = mean(GC4),
            median_GC4 = median(GC4)), .id = 'name')
```


## Figure 4 - Organelle linkage

Data preparation:

```{r}
ld_dflist <- fs::dir_ls('data/organelle-linkage/', regexp = '\\.txt') %>%
  map_dfr(read_delim, delim = ' ', col_types = cols(), .id = 'name') %>% 
  mutate(name = str_replace(name, 'data/organelle-linkage/', '')) %>% 
  mutate(name = str_replace(name, '.txt', '')) %>% 
  split(.$name)
```

Mean D':

```{r}
ld_dflist %>% 
  map(~ summarise(., mean_dprime = mean(dprime))) %>% 
  unlist()
# allpairs = 0.781
# minus = 1.0
# mt only = 1.0
# plus = 0.611
```

Plots:

```{r}
linkage_plot_theme <- function(font_size = 16) {
  theme(axis.title = element_text(family = "Helvetica", size = font_size),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        axis.text = element_text(family = "Helvetica", size = font_size, color = 'black'),
        axis.line = element_line(colour = 'black', linetype = 'solid'),
        axis.line.x = element_line(size = 0.9),
        axis.line.y = element_line(size = 0.9),
        axis.ticks = element_blank(),
        plot.tag = element_text(colour = 'black', size = font_size, face = 'bold'),
        panel.background = element_blank())
}

cp_plot <- ggplot(ld_dflist$plus, aes(x = dprime)) +
  geom_histogram(binwidth = 0.1, fill = wes_palette('Darjeeling1')[2]) +
  linkage_plot_theme(16) +
  coord_cartesian(x = c(-0.05, 1.05)) +
  labs(x = "D'", y = 'count', tag = 'a') +
  annotate_textp(
    'text', x = 0.05, y = 0.9, size = 16, 
     label = "cpDNA x MT+")

mt_plot <- ggplot(ld_dflist$minus, aes(x = dprime)) +
  geom_histogram(binwidth = 0.1, fill = wes_palette('Darjeeling1')[3]) +
  linkage_plot_theme(16) +
  coord_cartesian(x = c(-0.05, 1.05)) +
  labs(x = "D'", y = 'count', tag = 'b') +
  annotate_textp(
    'text', x = 0.05, y = 0.9, size = 16,
    label = 'mtDNA x MT-'
  )

mt_only_plot <- ggplot(ld_dflist$mt_only, aes(x = dprime)) +
  geom_histogram(binwidth = 0.1, fill = wes_palette('Darjeeling1')[4]) +
  linkage_plot_theme(16) +
  coord_cartesian(x = c(-0.05, 1.05)) +
  labs(x = "D'", y = 'count', tag = 'c') +
  annotate_textp(
    'text', x = 0.05, y = 0.9, size = 16,
    label = 'mtDNA'
  )

allpairs_plot <- ggplot(ld_dflist$allpairs, aes(x = dprime)) +
  geom_histogram(binwidth = 0.1, fill = wes_palette('Darjeeling1')[5]) +
  linkage_plot_theme(16) +
  labs(x = "D'", y = expression(paste('count (x', 10^3, ')')), tag = 'd') +
  coord_cartesian(x = c(-0.05, 1.05)) +
  scale_y_continuous(labels = seq(0, 400, by = 100)) +
  annotate_textp(
    'text', x = 0.05, y = 0.9, size = 16,
    label = 'autosomal'
  )

# combining plots
fig_4 <- cp_plot + mt_plot + mt_only_plot + allpairs_plot
fig_4

ggsave('plots/fig_4.eps', plot = fig_4,
       width = par('din')[1] * 1.25, height = par('din')[1])

```

## Other stats

### Kruskal-Wallis tests

#### piN/piS over annotations

```{r}
piN_piS_all %>% 
  mutate(chrom = as.factor(chrom)) %>% 
  kruskal.test(piN_piS ~ chrom, data = .) 

pairwise.wilcox.test(piN_piS_all$piN_piS, as.factor(piN_piS_all$chrom), 
                     p.adjust.method = 'BH')
```

#### GC content over annotations

```{r}
gc_all %>% 
  mutate(chrom = as.factor(chrom)) %>% 
  kruskal.test(GC4 ~ chrom, data = .)

pairwise.wilcox.test(gc_all$GC4, as.factor(gc_all$chrom), 
                     p.adjust.method = 'BH')

leveneTest(GC4 ~ chrom, data = gc_all)
```


# Supplementary

## Fig. S1 - autosomal ZnS across the genome

Data preparation:

```{r}
autosomal_fnames <- dir_ls('data/ld-windowed/autosomal/')
autosomal_zns <- map_dfr(autosomal_fnames, read_delim, delim = ' ', col_types = cols(), .id = 'name') %>% 
  mutate(chrom = str_extract(name, 'chromosome_[0-9]{1,2}')) %>% 
  select(-name) %>% 
  select(chrom, everything()) %>%  # reorder
  filter(site_count >= 30) %>% 
  mutate(midpoint = start + 500)
```

Individual plots:

```{r}
if (!is_dir('plots/autosomal')) {
  dir_create('plots/autosomal/')
}

chrom_plot <- function(d) {
  plot_name <- d$chrom[1] # sloppy...
  d_plot <- ggplot(d, aes(x = midpoint, y = zns)) +
    geom_point(size = 1.2) +
    labs(
      x = 'position (bp)',
      y = expression(paste(Z[nS]))
    ) +
    zns_plot_theme(16) +
    ggtitle(plot_name)
  
  file_name <- paste0('plots/autosomal/', plot_name, '.pdf')
  ggsave(file_name, plot = d_plot,
         width = par('din')[1] * 1.25, height = par('din')[1])
}

autosomal_zns %>% 
  split(.$chrom) %>% 
  walk(~ chrom_plot(.)) # walk is silent map
```

Are the ZnS profiles of any of the chrs different from one another?

```{r}
# mean + SE
autosomal_zns %>% 
  mutate(chrom = str_extract(chrom, '[0-9]{1,2}')) %>% 
  mutate(chrom = factor(chrom, levels = as.character(seq(1, 17)))) %>% 
  group_by(chrom) %>% 
  summarise(mean_zns = mean(zns),
            sd_zns = sd(zns),
            n = n()) %>% 
  mutate(se_zns = sd_zns / sqrt(n)) %>% 
  ggplot(aes(x = chrom, y = mean_zns,
             ymin = mean_zns - se_zns,
             ymax = mean_zns + se_zns)) +
  geom_bar(stat = 'identity', fill = 'white', color = 'black') +
  geom_errorbar(size = 0.3, width = 0.5) +
  theme_classic() +
  NULL

# ridgeplot
autosomal_zns_ridgeplot <- autosomal_zns %>% 
  mutate(chrom = str_extract(chrom, '[0-9]{1,2}')) %>% 
  mutate(chrom = factor(chrom, levels = as.character(seq(1, 17)))) %>% 
  ggplot(aes(x = zns, y = chrom)) +
  geom_density_ridges2(
    rel_min_height = 0.01,
    scale = 1.2,
    quantile_lines = TRUE,
    quantiles = 2 # dividing line in middle is median
    ) +
  theme_classic() +
  theme(axis.text = element_text(family = 'Helvetica', color = 'black')) +
  labs(
    x = expression(Z[nS]),
    y = 'chromosome'
  )

autosomal_zns_ridgeplot

# regenerate this w/o monitors later
ggsave('plots/fig_S1.pdf', plot = autosomal_zns_ridgeplot,
       width = par('din')[1] * 0.75, height = par('din')[1] * 0.5)
  
```

Mean ZnS across all autosomes:

```{r}
autosomal_zns %>% 
  summarise(mean_zns = mean(zns)) # 0.34
```


## Fig. S2 - domain separated linear fits of Fst and Zns

Data preparation in Figure 2 section above

```{r}
fst_annotations <- data.frame(
  r2 = c(
    'italic(R) ^ 2 == 0.389 ~~ italic(p) == 6.36 %*% 10 ^ -9', 
    'italic(R) ^ 2 == 0.799 ~~ italic(p) < 2.2 %*% 10 ^ -16', 
    'italic(R) ^ 2 == 0.005 ~~ italic(p) == 0.599'),
  domain_f = factor(c('T', 'R', 'C'), levels = c('T', 'R', 'C')),
  x = c(0.5, 0.5, 0.5),
  y = c(1, 1, 1)
)

windowed_fst_plot <- windowed_filtered %>% 
  mutate(domain_f = factor(domain, levels = c('T', 'R', 'C'))) %>% 
  ggplot(aes(x = zns, y = Fst)) +
  geom_point() +
  geom_smooth(method = 'lm', se = FALSE) +
  labs(
    x = expression(Z[nS]),
    y = expression(F[ST])
  ) +
  coord_cartesian(
    x = c(0, 1),
    y = c(0, 1)
  ) +
  ylim(0, 1) +
  corr_plot_theme(16) +
  scale_x_continuous(labels = c(0, 0.25, 0.5, 0.75, 1)) +
  facet_wrap(vars(domain_f), nrow = 1) +
  geom_text(
    data = fst_annotations,
    mapping = aes(x = x, y = y, label = r2),
    parse = TRUE, size = 3
  )

windowed_fst_plot

ggsave('plots/fig_S2.pdf', plot = windowed_fst_plot,
       width = par('din')[1] * 0.75, height = par('din')[1] * 0.75)
```


## Fig. S3 - differences in ZnS across domains

Data preparation in Figure 2 section above

Estimated marginal means:

```{r}
zns_lsmeans <- emmeans::lsmeans(zns_fit, pairwise ~ domain)

zns_means <- as_tibble(zns_lsmeans$lsmeans)
```

Plot:

```{r}
zns_means_plot <- zns_means %>% 
  mutate(domain = factor(domain, levels = c('T', 'R', 'C'))) %>% 
  ggplot(aes(x = domain, y = lsmean,
             ymin = lsmean - SE, ymax = lsmean + SE)) +
  geom_point(size = 1.5) +
  geom_errorbar(width = 0.2, size = 0.4) +
  corr_plot_theme(16) +
  coord_cartesian(y = c(0, 1)) +
  labs(y = expression(Z[nS]))

zns_means_plot
  
ggsave('plots/fig_S3.pdf', plot = zns_means_plot,
       width = par('din')[1] * 0.75, height = par('din')[1] * 0.75)
```

## Fig. S4 - Boxplots of Fst b/w autosomal and MT locus

```{r}
mt_autosomal_fst <- polymorphism_shared %>% 
  select(Fst) %>% 
  bind_rows(
    select(polymorphism_other, Fst),
    .id = 'annotation') %>% 
  mutate(
    annotation = case_when(
      annotation == 1 ~ 'MT',
      annotation == 2 ~ 'autosomal'
    )
  )

fst_plot <- mt_autosomal_fst %>%
  ggplot(aes(x = annotation, y = Fst)) +
  geom_boxplot() +
  geom_jitter(width = 0.2) +
  corr_plot_theme(16) +
  labs(y = expression(F[ST])) +
  coord_cartesian(y = c(0, 1)) +
  stat_compare_means(method = 'wilcox')

fst_plot

mt_fst <- mt_autosomal_fst %>% 
  filter(annotation == 'MT')
autosomal_fst <- mt_autosomal_fst %>% 
  filter(annotation == 'autosomal')

wilcox.test(mt_fst$Fst, autosomal_fst$Fst, alternative = 'two.sided')

ggsave('plots/fig_S4.pdf', plot = fst_plot,
       width = par('din')[1] * 0.75, height = par('din')[1] * 0.75)

```


# Post-review additions

## MT locus plot with annotations

Modified from Fig. 1 code above

```{r}
gff <- read_tsv('data/references/final.strict.GFF3', 
                col_names = FALSE) %>% 
  filter(X1 == 'chromosome_6') %>% 
  select(
    chrom = X1,
    type = X3,
    start = X4,
    end = X5,
    strand = X7,
    info = X9
  ) %>% 
  filter(type == 'mRNA', start >= 298298, end <= 943474) %>% 
  mutate(name = str_extract(info, 'Name\\=([A-Za-z0-9\\.\\:]+)\\;')) %>% 
  mutate(name = str_replace(name, 'Name\\=', '')) %>% 
  mutate(name = str_replace(name, ';', '')) %>% 
  mutate(gene_name = str_extract(info, 'geneName\\=[A-Za-z0-9\\.]+;')) %>% 
  mutate(gene_name = str_replace(gene_name, 'geneName\\=', '')) %>% 
  mutate(gene_name = str_replace(gene_name, ';', ''))
  # tidyr::unite(range, start, end, sep = '-', remove = TRUE)
  

View(gff)

# fig 1 replicate
zns_plot_markers <- plot_data %>%
  ggplot() +
  geom_rect(aes(xmin = 298000, xmax = 420000, ymin = -0.5, ymax = 1.5),
            color = '#eaecef', fill = '#eaecef') +
  geom_rect(aes(xmin = 421000, xmax = 826000, ymin = -0.5, ymax = 1.5),
            color = '#d9dce0', fill = '#d9dce0') +
  geom_rect(aes(xmin = 827000, xmax = 943000, ymin = -0.5, ymax = 1.5), 
            color = '#eaecef', fill = '#eaecef') +
  geom_point(aes(x = midpoint, y = zns, color = state), size = 1.2) +
  labs(
    x = 'position on chromosome 6 (bp)',
    y = expression(paste(Z[nS]))
  ) +
  scale_color_manual(
    values = c(hmm_colors[2], hmm_colors[1], hmm_colors[3])) +
  scale_x_continuous(
    breaks = seq(0, 40, by = 5)*1e5, 
    labels = scales::comma,
    limits = c(0, 31.5)*1e5) +
  coord_cartesian(
    x = c(0, 30.5)*1e5, y = c(0, 1)) +
  zns_plot_theme(16) +
  annotate('text', label = 'T', x = 360000, y = 0.05, size = 8) +
  annotate('text', label = 'R', x = 638000, y = 0.05, size = 8) +
  annotate('text', label = 'C', x = 884000, y = 0.05, size = 8) +
  guides(color = FALSE) +
  geom_rect(data = gff,
            aes(xmin = start, xmax = end,
                ymin = -0.04, ymax = -0.015,
                fill = strand)) +
  scale_fill_viridis_d() +
  guides(fill = FALSE) +
  NULL

zns_plot_markers

# mt zoomed in
zns_plot_mt <- plot_data %>% 
  ggplot() +
  geom_rect(aes(xmin = 298000, xmax = 420000, ymin = -0.5, ymax = 1.5),
            color = '#eaecef', fill = '#eaecef') +
  geom_rect(aes(xmin = 421000, xmax = 826000, ymin = -0.5, ymax = 1.5),
            color = '#d9dce0', fill = '#d9dce0') +
  geom_rect(aes(xmin = 827000, xmax = 943000, ymin = -0.5, ymax = 1.5), 
            color = '#eaecef', fill = '#eaecef') +
  geom_point(aes(x = midpoint, y = zns, color = state), size = 1.2) +
  labs(
    x = 'position on chromosome 6 (bp)',
    y = expression(paste(Z[nS]))
  ) +
  scale_x_continuous(
    labels = scales::comma
  ) +
  scale_color_manual(
    values = c(hmm_colors[2], hmm_colors[1], hmm_colors[3])) +
  ggplot2::coord_cartesian(
    x = c(280000, 950000), y = c(0, 1)) +
  zns_plot_theme(16) +
  annotate('text', label = 'T', x = 360000, y = 0.05, size = 8) +
  annotate('text', label = 'R', x = 638000, y = 0.05, size = 8) +
  annotate('text', label = 'C', x = 884000, y = 0.05, size = 8) +
  guides(color = FALSE) +
  geom_rect(data = gff, 
            aes(xmin = start, xmax = end,
                ymin = -0.04, ymax = -0.015,
                fill = strand)) +
  scale_fill_viridis_d() +
  guides(fill = FALSE) +
  NULL

zns_plot_mt

```


Saving the plots:

```{r}
ggsave('plots/fig_1_with_markers.pdf', plot = zns_plot_markers,
       width = par('din')[1] * 1.50, height = par('din')[1] * 0.85)

ggsave('plots/mt_locus_markers.pdf', plot = zns_plot_mt,
       width = par('din')[1] * 1.50, height = par('din')[1] * 0.85)
```

## Table S2 - differentiation and diversity across MT

```{r}
# polymorphism_shared loaded in at fig. 2 above
table_s2 <- polymorphism_shared %>% 
  select(
    gene, Fst, zns = zns_all, GC4,
    shared, fixed, contains('piN_piS'), 
    contains('variant'), contains('private')
  )

write_tsv(table_s2, path = 'data/cds-popgen/table_s2.tsv')
```








