
## `alignment-lastz`

Scripts that use the output of a LASTZ alignment (this work uses an HSP threshold of
30k) between the mt alleles to create FASTA files for downstream analysis.

From the LASTZ alignment, both `.maf` and `general` (referred to as `bed`) files
are generated. These contain equivalent data encoded different ways. 
The `maf` file has information about where gaps are in the alignment, while
the `general` file is easier to parse for other purposes. 

### Filtering alignments - `clean_lastz_output.R`

This script filters a `general` formatted LASTZ output file
to resolve overlapping alignments by only keeping the alignment with
the highest score.

### Creating aligned FASTA files - `align_mt_fasta_maf.py`

This script uses a `.maf` formatted LASTZ output file alongside
the filtered `general` file generated by `clean_lastz_output.R` to
create a directory of FASTA alignments across both mating types. Each file 
in the directory represents a single alignment in the LASTZ output.

### Combining the FASTAs into a single 'master' alignment

This task involves three scripts.

`transpose_fastas.py` transposes the generated alignments into a
'long' format, with positions as keys and nucleotides as values.
This allows all the alignments to be combined into one file.

`remove_duplicates.R` reduces the overlap between some alignments
to a single instance of each positions.

`combine_fastas.py` converts the filtered, transposed sequence file
back into FASTA format. The alignment is as long as the entire mt+ locus,
and non-aligned (ie non gametologous) regions are filled in with Ns.

### Creating non-gametolog files for the mt alleles - `make_mt_only.py`

This script uses the LASTZ alignment to create FASTA files with the
_gametologous_ regions masked. It is effectively the inverse of the
file created by `combine_fastas.py` above. The files generated by this
script are to be used for other population genetic analyses if need be,
but NOT for recombination estimation, since the masking of gametologous
regions otherwise inflates the per bp recombination rates in these regions.
